{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bdc25a86",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "USER_AGENT environment variable not set, consider setting it to identify your requests.\n",
      "/home/bionik/miniconda3/envs/LLM/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import faiss\n",
    "from langchain_ollama import ChatOllama,OllamaEmbeddings\n",
    "from langchain_core.prompts import PromptTemplate\n",
    "from pydantic import BaseModel, Field\n",
    "from langchain_community.document_loaders import WebBaseLoader, PyPDFLoader, TextLoader\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n",
    "from langchain_community.vectorstores import FAISS\n",
    "import faiss\n",
    "import pyttsx3\n",
    "from langchain_core.output_parsers import PydanticOutputParser,StrOutputParser\n",
    "from typing import List\n",
    "import re\n",
    "from langchain_core.runnables import RunnableParallel, RunnablePassthrough\n",
    "import numpy as np\n",
    "from langchain_core.runnables import RunnablePassthrough\n",
    "from langchain_core.prompts import ChatPromptTemplate,SystemMessagePromptTemplate,HumanMessagePromptTemplate\n",
    "from langchain_core.prompts import PromptTemplate\n",
    "import spacy\n",
    "import pickle\n",
    "import requests\n",
    "from scispacy.linking import EntityLinker\n",
    "import scispacy\n",
    "import re\n",
    "from langchain.messages import SystemMessage,HumanMessage,AIMessage,ToolMessage\n",
    "from langchain.tools import tool\n",
    "from langchain_community.docstore.in_memory import InMemoryDocstore\n",
    "import pyobo\n",
    "import time\n",
    "from collections import defaultdict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97dbbc35",
   "metadata": {},
   "outputs": [],
   "source": [
    "#linker = pyobo.get_scispacy_entity_linker(\"uniprot\", filter_for_definitions=False, resolve_abbreviations=True)\n",
    "nlp = spacy.load('en_ner_jnlpba_md')\n",
    "nlp.disable_pipes(\"tagger\", \"parser\")\n",
    "nlp.add_pipe(\"sentencizer\")\n",
    "linker = pyobo.get_scispacy_entity_linker(\"hgnc\", filter_for_definitions=False, resolve_abbreviations=True,)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "03fe835f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_base_url='https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi'\n",
    "query='\"signaling pathway\"[Title/Abstract] AND JAK/STAT AND review[Publication Type] AND (humans[MeSH Terms] AND (\"disease\"[MeSH Terms] OR disease[Title/Abstract]))'\n",
    "params_id = {\"db\": \"pubmed\",\"term\": query,\"retmode\": \"json\",\"retstart\": 0,\"retmax\": 10000}\n",
    "response=requests.get(ncbi_base_url, params=params_id)\n",
    "pubmed_id=response.json()['esearchresult']['idlist']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b71e9b17",
   "metadata": {},
   "outputs": [],
   "source": [
    "pub_2_pmc={}\n",
    "pubmed_2_PMC_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi\"\n",
    "for i in pubmed_id:\n",
    "    params_PMC = {\"dbfrom\": \"pubmed\",\"db\": \"pmc\",\"id\": i,\"retmode\": \"json\"}\n",
    "    response = requests.get(pubmed_2_PMC_url, params=params_PMC)\n",
    "    id=response.json()['linksets'][0]\n",
    "    if 'linksetdbs' in id.keys():\n",
    "        pub_2_pmc[i]=int(id['linksetdbs'][0]['links'][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "df141d83",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_docs=[]\n",
    "for i in pubmed_id: \n",
    "    if i in pub_2_pmc.keys():\n",
    "        doc=WebBaseLoader(f'https://pmc.ncbi.nlm.nih.gov/articles/{pub_2_pmc.get(i)}/').load()\n",
    "        all_docs.extend(doc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "2409f221",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Interstitial lung disease (ILD) is a significant extra-articular complication of rheumatoid arthritis (RA), characterized by high prevalence and mortality rates. Although advancements have been made in understanding its potential mechanisms, the pathogenesis of RA-associated ILD remains incompletely understood. Recent research has shed light on roles of various disease-related signaling pathways, including TGF-β/SMAD, JAK/STAT, PI3K–Akt, Wnt/β-catenin, and NF-κB, which are implicated in development of both RA and lung fibrosis. These shared pathways, which drive CCL1 PROTEIN production and fibroblast proliferation, offer promising opportunities for therapeutic intervention, including pathway-specific inhibition and drug repurposing. Furthermore, the growing identification of potential biomarkers for early detection and severity assessment in RA-ILD patients holds promise for improving clinical management and guiding treatment strategies. Current treatments fall short in effectively halting the progression of lung fibrosis. This highlights the potential of advancements in signaling pathways and targeted therapies as promising alternatives with significant opportunities for improvement. : rheumatoid arthritis, interstitial lung disease, pathogenesis, signaling pathway, targeted therapy1. Introduction Rheumatoid arthritis (RA) is a chronic systemic autoimmune disease characterized by symmetric polyarthritis which can cause irreversible joint damage and significant disability . In addition to the common articular involvement, extra-articular manifestations can also occur in RA patients, contributing to overall morbidity and mortality . The lungs, in particular, are frequently affected in individuals with RA experiencing lung-related issues . The disease can affect the lung’s interstitium, airways, and pleurae . Among these, interstitial lung disease (ILD) stands out due to its potential to cause irreversible pulmonary fibrosis, which can significantly worsen the prognosis for those affected, underscoring the importance of comprehensive care in managing RA . The histological features of RA-ILD can be categorized into usual interstitial pneumonia (UIP), nonspecific interstitial pneumonia (NSIP), organizing pneumonia (OP), desquamative interstitial pneumonia (DIP), and lymphocytic interstitial pneumonia (LIP) . Different from most forms of connective-tissue-associated ILD, UIP is more prevalent than NSIP, occurring in up to 61% of RA-ILD patients (7, 8). It shows poor responses to both glucocorticoids and immunosuppression . Therefore, patients with progressive fibrosing phenotypes are restricted to antifibrotic drugs, which often leads to high mortality . Knowledge regarding mechanisms of RA-ILD has been greatly improved in recent years, as accumulated experimental evidence allows more detailed elucidation of potential molecular mechanisms and immunological pathways (4, 11, 12). However, many pathophysiological connections remain unclear. Based on knowledge of current literature, the central need in this area is for early recognition to prevent disease progression and effective therapeutic alternatives to reverse the condition (13, 14). However, this requires clear insights into pathogenesis and a better understanding of shared pathways involved. This review provides an overview of the key mechanisms in the pathogenesis of ILD in RA, as well as the advances in clinical screening and targeted therapies, highlighting areas of shared immunological pathways and ongoing uncertainties in related therapeutic opportunities.2. Pathogenesis The pathogenesis of RA-ILD usually involves the participation of multiple risk factors. Genetic predisposition has been implicated in the development of fibrosis and susceptibility to RA-ILD . Continuous epithelial damage, cellular interactions, and the interplay between the various inflammatory components could lead to the activation of fibroblasts and extracellular matrix deposition, which probably determines the progression of pulmonary fibrosis . During these processes, multiple signaling transduction pathways, including TGF-β/SMAD, JAK/STAT, PI3K–Akt, Wnt/β-catenin, and NF-κB signaling pathways were activated, triggering biological effects of cell proliferation and protein synthesis . 2.1. Genetic predisposition and risk factors Potential risk factors related to the pathogenesis of RA-ILD are summarized in. Both endogenous and exogenous risk factors can contribute to RA-ILD. Previous genetic studies have identified one specific variant of MUC5B DNA that serves as a strong indicator of increased risk of IPF . Specifically, this variant has been suggested to be associated with RA-ILD patients of the UIP pattern . For genes concerned with telomere maintenance, one study of whole-exome sequencing exhibited higher mutation frequency of TERT, PARN PROTEIN and RTEL1 PROTEIN in RA-ILD, which was previously related to familial pulmonary fibrosis . Patients with a TERT, RTEL1 DNA or PARN PROTEIN mutation have shorter telomeres compared with control groups, indicating that premature senescence may play a role in fibrosis and lung remodeling in RA patients . Genetic predisposition of HLA genes to ILD in RA patients has been studied. A lower risk of ILD was found in RA population with shared epitope of HLA-DRB1 DNA in a Japanese cohort . Differently, HLA-DRB1 PROTEIN and HLA-DRB1 PROTEIN from the HLA-DR2 alleles were correlated with a higher risk of ILD . Additionally, another genetic variant at RPA3-UMAD1 was identified in Japanese individuals with RA-ILD .Summary of risk factors related to the pathophysiology of RA-ILD. Both endogenous and exogenous risk factors can contribute to the development of RA-ILD. Individuals with a genetic predisposition, especially older males, are at higher risk of developing RA-ILD. Environmental exposure to pollutants, cigarette smoking, and pro-inflammatory agents are also relevant. Internal dysregulations include epithelial damage, fibroblast activation, inflammation, oxidative stress, ACPA, and RF. RF, rheumatoid factor; ACPA, anti-cyclic citrullinated peptides antibody.Currently, risk factors associated with RA-ILD include genetic variations, older age, male gender, race, and cigarette smoking (23, 24). The abnormal immune response induced by chemical components of cigarettes may underlie the potential mechanism contributing to ILD formation . Meanwhile, autoantibodies, including RF and PRTN3 PROTEIN have been demonstrated to be associated with RA-ILD progression . Continuous exposure to risk factors contributes to epithelial deterioration and activation of immune cells, leading to the excessive accumulation of extracellular matrix in the lung .2.2. Epithelial damage, inflammation, and fibroblast activation Most theories propose the mucosal sites, including airway, gastrointestinal, or genitourinary tract, may serve as potential origins of inflammation . Among them, the airway mucosa plays a critical role in the initiation of RA-related autoimmunity, as it is the site where the predisposing environmental factors interact with the host, resulting in local injury and triggering an inflammatory response . Constant alveolar epithelial injury and immune stimulation induced by RA trigger inflammatory responses in pulmonary tissue, resulting in infiltration of inflammatory cells into the interstitial and alveolar airspaces . Their activation and interactions, along with the participation of a wide profile of cytokines, are critical to the generation of abnormal fibroproliferative responses, as summarized in.Possible mechanisms of rheumatoid arthritis (A) and related interstitial lung disease (B). Pathogenesis of parenchymal lung involvement in RA includes participation and activation of diverse immune cells such as T cells, B cells, macrophages, neutrophils, and eosinophils. They contribute to joint damage and pulmonary fibrotic responses through the secretion of pro-inflammatory cytokines. Fibroblasts both at affected sites and circulation are involved in the self-sustaining of lung fibrosis. APC CELL_TYPE, antigen-presenting cells; IFN-γ, interferon-γ; TNF-α, RELT PROTEIN; CXCL5 PROTEIN, C-X-C motif ligand; MMP9 PROTEIN, MMP19 PROTEIN; TNFSF11 PROTEIN, receptor activator of nuclear factor kappa-B ligand; EMT, epithelial-mesenchymal transition; FMT, fibroblast-to-myofibroblast transition; IL, CXCL8 PROTEIN; TGF-β, transforming growth factor-β; ECM, extracellular matrix; VEGF, VEGFA PROTEIN; PDGFC PROTEIN, PDGFA PROTEIN.One critical step in the development of interstitial fibrosis is related to activated fibroblasts and myofibroblasts resident within the lung, as well as recruitment of progenitor cells from blood circulation . Evidence from animal models suggested the localization and differentiation of circulating progenitor cells into fibrocytes, including mesenchymal stem cells and a monocyte lineage, contributing to the accumulation of SPON2 PROTEIN and tissue fibrosis . Additionally, the epithelial cells of lung tissue have been shown to participate in the profibrotic process via a mechanism known as epithelial–mesenchymal-transdifferentiation (EMT), during which they acquire mesenchymal characteristics and behaviors . This transition is upregulated in animal models of RA-ILD triggered by signals from the microenvironment, resulting in an increased profibrotic mesenchymal cell population (32, 33). Furthermore, fibroblast-myofibroblast transition (FMT) will then accelerate the process of fibrosis by translating activated fibroblasts into myofibroblasts, establishing a self-sustaining cycle .3. Activation of signaling pathways involved in RA and ILD 3.1. TGF-β/SMAD signaling pathway TGF-β signaling is an important pathway in fibrosis mechanism regulating response of tissue repair . Pleiotropic effects are mediated via three different types of TGF-β receptors, including TGFβRI, TGFβRII, and TGFβRIII . Dysregulation of TGF-β signaling has been implicated in disease development . In particular, over-activation of TGFβRI and TGFβRII can induce excessive COL5A2 PROTEIN accumulation and progressive fibrosis in pulmonary tissues (38, 39). Recent studies have revealed that the activation of TGF-β-SMAD2/3 signaling promotes EMT in lung fibrosis through establishing an animal model of RA-ILD mice . Furthermore, in another RA mouse model, pulmonary fibrosis can be reduced by inhibiting pathway of TGF-β/Smads signaling, during which proliferation and migration of RA synovial fibroblasts were also restricted . Insights from mechanism studies of autophagy also indicated that a decreased number of activated myofibroblasts can be induced by enhancing autophagy activity via inhibition of TGF-β1-Smad3/ERK/P38 signaling pathway . TGF-β inhibition therapy carries great expectations and are under investigation to improve safety and efficacy .Model of shared mechanistic pathways involved in both RA and ILD. Activation of TGF-β/SMAD, JAK/STAT, PI3K–Akt, Wnt/β-catenin, and NF-κB signaling pathways and resultant downstream reactions contribute to multiple biological process. Improved understanding of their roles in RA-ILD pathogenesis offers promising therapeutic targets. PIP PROTEIN, phosphatidylinositol bisphosphate; mTORC1, mechanistic target of rapamycin complex 1; STAT, STAT1 PROTEIN; TCF/LEF, LEF1 PROTEIN; GSK3A PROTEIN, NEK3 PROTEIN; EMT, epithelial-mesenchymal transition.3.2. JAK/STAT signaling pathway JAK/STAT pathway can be activated by various cytokines and growth factors associated with fibrosis and inflammation, such as IL11 PROTEIN, IL6 PROTEIN, TGF-β1, and FGF12 PROTEIN (FGFs) (43, 44). It participates and mediates extensive biological processes, including cell differentiation and immune regulation, closely related to autoimmune disorders and fibrosis . The JAK/STAT signaling system is widely recognized as a key molecular mechanism in the etiology and development of RA, and it has emerged as a novel effective target in RA therapeutic strategies (46, 47). Given its involvement in several fibrotic disorders, there has been an interest in exploring its potential roles in ILDs . Several studies revealed that JAK2/STAT3 is the most common subtype in ILD, which participates in intercellular signal transduction in response to the stimulation of cytokines and growth factors (49, 50). RNA sequencing and immunohistostaining have provided evidence that JAK2 PROTEIN serves as a regulator and intermediator of TGF-β-mediated lung fibroblast responses in the development of RA-ILD with a usual interstitial pneumonia (RA-UIP) and IPF . In contrast, their findings suggested STAT5A PROTEIN as the downstream target in transcription, necessitating further investigation for clarification .3.3. PI3K–AKT1 PROTEIN signaling pathway The signaling pathway of PI3K-Akt has a great effect on cell activation and migration . In patients with ILD disease, it has been demonstrated to be aberrantly upregulated, contributing to tissue repair and disease progression . Enhanced activity of PI3K/AKT signaling pathway is related to synovial inflammation and osteoclasts proliferation in RA pathogenesis . PI3Kγ affects innate immunity of antigen-induced arthritis in the early phase by regulating downstream signal molecules, during which activated macrophage and neutrophil infiltrate into the joint . Research on PI3Kγ has also indicated that it may regulate synovial inflammation via fibroblasts, resulting in cartilage destruction . Impact of PI3K/AKT signaling pathway on osteoclast differentiation and generation may also contribute to aggravation of RA . Consequently, PIK3CA PROTEIN represents a viable therapeutic target for RA due to its central role in mediating inflammatory responses and bone destruction. Aberrant activation of PI3K/AKT has also been observed in IPF patients and demonstrated to play a central role in pathogenesis of the disease, which was considered as an effective therapeutic target against lung fibrosis (57, 58). Our research team has recently found that the CXCL16/CXCR axis can facilitate pulmonary fibrosis by activating PI3K/AKT/FOXO3a signaling pathway in human pulmonary fibroblasts, stimulating cell proliferation and COL5A2 PROTEIN production . Another study elucidated upregulation of PI3K-Akt signaling pathways in RA-ILD, based on network pharmacological and metabolic perspectives . Pharmacological inhibitors targeting PI3K/AKT pathway are presently being investigated in clinical trials for IPF treatment, demonstrating promising potential as innovative anti-fibrotic therapeutics .3.4. Wnt/β-catenin signaling pathway Research on Wnt pathway in RA has suggested that it plays a role in synovial cell proliferation and bone homeostasis, promoting synovial hyperplasia and contributing to joint damage (61, 62). The Wnt/β-catenin signaling pathway was significantly activated in IPF, with participation of a sequence of related ligands and receptors . Molecules involved in Wnt pathway, such as CCNO PROTEIN and MMP7 PROTEIN, were found to be over-expressed in lesions of pulmonary fibrosis . One study focusing on M2 macrophages indicated that inhibition against Wnt/β-catenin signaling pathway may suppress myofibroblast differentiation and pulmonary fibrosis (65), suggesting that repurposing Wnt-targeting agents could hold promise for RA-ILD. Since shared Wnt signaling was involved in ILD and RA, a recent study investigated the clinical significance of a representative ligand Wnt5 in RA-ILD: Elevated levels of WNT5A PROTEIN were observed in patients with RA-ILD, especially in patients with UIP pattern . Accordingly, evaluation of circulating WNT5A PROTEIN may provide a potential candidate biomarker for early identification and progression assessment of RA-ILD . Furthermore, Wnt5a inhibitors or receptor antagonists could be explored to mitigate fibrotic progression in RA patients.3.5. NF-κB signaling pathway NF-κB signaling pathway is widely acknowledged for its role in the activation and differentiation of various immune cells, closely related to human inflammatory response . Dysregulated activation of NFYA PROTEIN promotes chronic inflammation and excessive proliferation of LERFS CELL_TYPE in RA (68, 69). Activated NF-κB members in the synovial tissue of RA patients can induce activation and differentiation of immune cells as well as the upregulation of various pro-inflammatory cytokines such as TNF-α, IFNA1 PROTEIN and IL17A PROTEIN, contributing to the aggregation of RA progression . It was also found in an animal model of lung fibrosis that fibroblast-to-myofibroblast differentiation, which contributes to aberrantly activated lung myofibroblasts and extracellular matrix accumulation, is dependent on the activation of NF-κB signaling . Moreover, one study supports IL17A PROTEIN of its direct role in stimulating positive responses of lung fibroblasts proliferation and extracellular matrix accumulation in RA-ILD pathogenesis via signaling mediated by NF-κB . Therapeutically, this underscores why NF-κB inhibition remains a key strategy in RA management, while new approaches are being explored to directly target this pathway and mitigate lung fibrosis.4. Biomarkers of clinical identification and monitoring Biomarkers correlated with RA-ILD development serve as useful tools for early screening and intervention guidance. Autoantibodies, cytokines, and novel biomarkers of other properties provide valuable guidance in screening and prediction of disease progression. ( Table 1 ) As a group of auto-antibodies widely used for early RA diagnosis, PRTN3 PROTEIN is also identified as a risk factor and strong predictor of RA-ILD (73, 74). It may become a useful serological biomarker to predict the presence and progression of lung fibrosis in RA patients (75, 76). Additionally, an elevated level of RF is also suggested to be associated with clinically evident and subclinical RA-ILD risk (75, 77). Recent studies indicate that the autoantibodies targeting citrullinated Hsp90α and HSP90AA1 CELL_LINE in patients with RA-ILD exhibit moderate sensitivity and great specificity in predicting RA-ILD progression . Table 1. Potential serum biomarkers for RA-ILD. Biomarkers Property Potential value in clinical evaluation ',\n",
       " ' ACPA Autoantibody Elevated levels associated with clinically evident and subclinical RA-ILD; specific PRTN3 PROTEIN improve RA-ILD prediction (75, 76) Rheumatoid factors Autoantibody Elevated levels associated with clinically evident and subclinical RA-ILD; a potentially useful prognostic marker of disease activity and mortality risk (75, 77) Anti-citrullinated Hsp90α/β Autoantibody Distinguishing RA-ILD patients from RA without ILD, or MCTD, IPF patients; useful biomarkers of high specificity and moderate sensitivity (78) MMP-7 Extracellular remodeling CELA3B PROTEIN Elevated in RA patients with organ-specific manifestation of ILD; contribute to the identification of RA-ILD during subclinical stage; negatively correlated with PFT variables (75, 80, 160) CXCL10 PROTEIN Elevated in RA patients with organ-specific manifestation of ILD; a sensitive biomarker of disease activity and predictor of RA-ILD (80, 161) IL-13 Cytokine Detecting the severity of ILD in RA patients (162) IL18 PROTEIN Associated with RA-ILD diagnosis and progression of lung disease (163) CCL18 PROTEIN Higher concentrations in patients with RA-ILD (164, 165) PARC Cytokine Contribute to identification of RA-ILD during subclinical stage (75) sPD-L1 Soluble checkpoint inhibitor A predictive marker for detecting the presence of ILD (166) LOXL2 PROTEIN Oxidative stress marker Diagnostic value for RA patients with ILD duration less than 3 months (167) MUC1 PROTEIN Lung epithelium-derived protein Correlated with HRCT results and PFT variables; a circulating marker for disease diagnosis and progression (160, 165, 168) CHI3L1 PROTEIN Lung SERPINF1 PROTEIN useful biomarker in diagnosis and severity evaluation (161, 169) Periostin Matrix protein A potential biomarker for diagnosis and fibrosis evaluation (170) SP-D Lung epithelium-derived protein Contribute to the identification of RA-ILD during subclinical stage; negatively correlated with PFT variables (75, 160) Uric acid Metabolite A diagnostic biomarker in RA-ILD, particularly in RA-UIP (85) Hsa-miR-214-5p and has-miR-7-5p MicroRNA Useful biomarkers for diagnosis (88)  ACPA, anti-citrullinated peptide antibodies; MCTD, mixed connective tissue disease; MMP-7, MMP19 PROTEIN; PFT, Pulmonary Function Test; CXCL5 PROTEIN, CXCL2 PROTEIN; IL, Interleukin; CCL, chemoattractant cytokine ligand; CCL18 PROTEIN, pulmonary and activation-regulated chemokine; CD274 PROTEIN, soluble programmed death ligand-1; LOXL2 PROTEIN, LOXL1 PROTEIN like 2; MUC1 PROTEIN, Krebs von den Lungen-6; CHI3L1 PROTEIN, Chitinase-3-Like protein; SP-D, surfactant associated protein D.Matrix metalloproteinases (MMPs) are enzymes involved in extracellular matrix degradation and tissue remodeling . The serum concentration of MMP-7 and CXCL10 PROTEIN was elevated in patients with RA-ILD compared with RA without ILD . Meanwhile, another research manifested a profile of biomarkers relating to lung disease, including MMP-7, activation-regulation chemokines, and surfactant protein D, which may be conducive to early diagnosis during the subclinical stage (81, 82). One study compared the profiling of serum proteins and demonstrated that lysyl oxidase-like 2 (LOXL2 PROTEIN) levels were significantly increased in RA individuals, especially in those patients with ILD duration less than three months . Krebs von den Lungen-6 (MUC1 PROTEIN) is a MUC2 PROTEIN associated with pulmonary fibrosis and is also a promising indicator for RA-ILD diagnosis and severity assessment (83, 84). It is noteworthy that, in addition to protein, other chemicals can also provide valuable information in disease monitoring. Evidence from the latest studies indicated uric acid can become a useful diagnostic biomarker in RA-UIP (85–87). Interestingly, a noticeable positive connection between microRNA, especially Hsa-miR-214-5p and has-miR-7-5p, and RA-ILD was also discovered recently, which exhibits prospectives in future studies .5. Existing therapies and safety concerns According to treatment algorithms for RA-ILD recently built by the ERS/EULAR clinical practice guidelines, pirfenidone is recommended for patients with UIP pattern and combination of immunosuppressive treatment is suggested in patients diagnosed with severe or progressive ILD . Moreover, nintedanib or combination of immunosuppressant and nintedanib is recommended for individuals with progressive pulmonary fibrosis . For RA-ILD with active arthritis, managing RA to target is of great importance. Currently, drug therapy for RA, such as conventional DMARDs (cDMARDS) and biologic DMARDs (bDMARDs) is widely recognized, while such effects may be balanced by higher risks of drug-related pulmonary disease . Until now, no evidence from randomized controlled trials (RCTs) has been found to optimize RA-ILD patient management. ( Table 2 ). Table 2. Recent clinical trials of pharmacological interventions in RA-ILD. Intervention Trial name Phase Population Primary ending point Current status Pirfenidone TRAIL1(155) II RA-ILD Decline in FVC% from a baseline of 10% or more or death Completed DMARDs & pirfenidone NCT04928586 IV CTD-ILD Change in FVC and DLCo Active, not recruiting Pirfenidone & glucocorticoid & immunosuppressant NCT05505409 IV CTD-ILD Change in FVC% Recruiting Abatacept APRIL II RA-ILD Change of FVC Unknown status Nintedanib NCT05503030 Observational study CTD-ILD Change in FVC and dyspnoea symptom score Recruiting Tofacitinib RAILDTo II RA-ILD Incidence and severity of adverse events Recruiting Tofacitinib vs Methotrexate PULMORA IV RA-ILD Change in total score by HRCT Terminated Genakumab injection NCT06189495 II RA-ILD Change in FVC, DLCo, PGA, and adverse events Not yet recruiting N-acetylcysteine NCT01424033 II, III CTD-ILD Pulmonary Function Tests Terminated  DMARDs, disease-modifying anti-rheumatic drugs; CTD-ILD, connective tissue disease-associated interstitial lung disease; FVC, forced vital capacity; DLCo, diffusing capacity for carbon monoxide; HRCT, high resolution CT; PGA, patient global assessment.5.1. Methotrexate Methotrexate (MTX) is the anchor DMARD in treating RA patients . However, despite its significant efficacy and tolerability, a range of potential side effects, especially pulmonary toxicity, cannot be overlooked . It is conventionally believed that MTX is associated with a higher risk of adverse respiratory events, including fibrotic ILD . According to some retrospective studies, MTX is correlated with the occurrence or exacerbation of ILD, constituting as much as 7.6% of the ILD incidence . On the other hand, the direct causality of MTX and ILD remains undefined. According to one meta-analysis in 2021, MTX treatment is not associated with a higher risk of RA-ILD among RA patients . One Korean prospective cohort of RA-ILD patients also exhibits similar results, indicating that MTX does not contribute to the increased risk of ILD progression . Even better, recent evidence from a multi-center prospective cohort study suggested MTX can reduce the onset of ILD, supporting the beneficial role of MTX in RA-ILD . In addition, recently research has suggested that use of MTX is related to improvement and longer survival among RA-ILD patients . By contrast, prospective studies can avoid time-window bias, thereby providing stronger evidence of causality. Therefore, traditional view of correlations between MTX and ILD progression has been challenged.5.2. Biologic drugs 5.2.1. TNF PROTEIN inhibitors Currently, no guidelines have been established when deciding biologic agents or targeted therapies in RA-ILD patients. The efficiency of anti-TNF agents in alleviating articular and synovial damage is supported by previous studies, while their role in stabilizing RA-ILD remains controversial (99, 100). Numerous studies have demonstrated that no significant beneficial effects were seen between patients with RA-ILD treated with anti-TNF agents and other therapies (101, 102). Moreover, it is also suggested that a higher prevalence of ILD among RA patients is associated with TNF-inhibiting treatment . Results of one cohort study illustrated a lower survival rate of RA-ILD patients using TNFi as the first-line therapy by comparing their 5-year mortality rates with groups taking rituximab (RTX) as the first-line treatment . Another prospective case-control study has also demonstrated the exacerbation of ILD in cases treated with TNFi compared with non-TNFi biologic agents including tocilizumab and abatacept. Additionally, TNFi is indicated to be related to a higher incidence or development of ILD according to their finding . Therefore, given the current understanding, increased focus is necessary to ensure the efficacy and safety of biological therapy in RA-ILD patients .5.2.2. Rituximab, abatacept and tocilizumab Other bDMARDs, such as RTX and abatacept are more advised to be considered as alternatives in treating RA-ILD. According to a ten-year retrospective observational cohort conducted among 700 RA patients, RTX can serve as an acceptable therapeutic option with efficacy in stabilizing RA-ILD . Significant improvement in %-predicted FVC was found in RA patients with RA-ILD under rituximab treatment . Meanwhile, evidence from systemic review suggested a significant lower worsening rate of ILD under therapy of abatacept in RA-ILD compared with TNF PROTEIN inhibitors, which therefore would become a desirable alternative for RA-ILD patients . Moreover, one recent cohort study on tocilizumab declared that positive effects were shown in most RA-ILD patients during a 6-month period by monitoring biomarkers of both MMP-3 and MUC1 PROTEIN. However, it is noteworthy that condition of ILD can be worsened by respiratory infection after one year of tocilizumab therapy . Currently, ongoing clinical trials on RA-ILD are summarized in Table 2 . Due to the absence of prospective randomized controlled studies for further confirmation, more detailed therapeutic guidelines and recommendations are yet to be established.6. Pathway-targeted therapy 6.1. Targeting JAK/STAT signaling pathway JAK inhibitors, such as tofacitinib and baricitinib, are novel effective therapies in the treatment of RA (111, 112). Based on data from the most recent studies, JAKi shows great tolerance in RA-ILD patients and comparable efficacy and safety to bDMARDs including RTX and abatacept . Since they can alleviate the joint damage of RA patients by controlling serum inflammatory cytokines and exerting anti-fibrotic effects, they may also lead to clinical improvement in RA patients with ILD . One retrospective study examined the effectiveness of JAK inhibitors in patients with RA-ILD, among whom 83.9% of patients showed clinical stability or improvement after 18 months of treatment . Meanwhile, studies in 2019 demonstrated the pathogenic role of IL17A PROTEIN, which is a T helper 17 cytokine, involved in the pathogenesis of RA and pulmonary fibrosis. Their observations have supported the fibrogenic response in RA-ILD induced by IL17A PROTEIN, which can be significantly alleviated with the inhibition of JAK2 PROTEIN . Despite limited knowledge from clinical trials, tofacitinib, a JAK inhibitor, represents an effective anti-inflammatory therapeutic option for RA patients with lung involvement. Case reports have suggested that tofacitinib contributes to the stabilization of ILD in RA patients (116, 117). Recently, a large retrospective cohort consisting of 28,559 RA patients treated with tofacitinib showed a much lower incidence of ILD compared with patients treated with other bDMARDs, suggesting their significance in reducing risk of ILD in RA . Baricitinib, a JAK1/JAK2 inhibitor, was demonstrated to show promising effects in RA patients with interstitial lung involvement. In patients under baricitinib therapy, significant improvement in lung diffusion index and reduction of biomarkers was observed . Additionally, administration of JAK1 PROTEIN selective inhibitor upadacitinib in RA-ILD patients who were resistant to MTX and TNFi has shown improved outcomes, as reported by retrospective studies as well as individual cases (119, 120). However, safety concerns with JAKi are also noteworthy. Black box warnings have already been raised by FDA, attaching importance to increased risks of cardiovascular diseases, infection and cancer, inducing regulatory restrictions . According to evidence from clinical trials, adverse outcomes are more commonly observed in the elderly, or individuals with baseline cardiovascular risks, inflammation status and disease activity . Further investigations into JAK inhibitors as novel therapeutic options to treat RA-ILD patients as well as their safety profile will be of great significance.6.2. Other promising pathway-related therapeutic options Since shared molecular pathways exist in both RA and ILD, key molecules involved in signaling transduction may provide promising targets for developing novel therapeutic strategies for RA-ILD patients. PDE12 PROTEIN (PDE) inhibitors were found to exert a potent effect in anti-remodeling via inhibiting the target of the TGF-β1 signaling pathway against lung fibrosis . Given that they possess potent anti-inflammatory activities by modulating various immune cell functions, they are also considered a potential therapeutic strategy for RA (124, 125). Additionally, selective PDE4 inhibitors may also inhibiting the inflammatory reaction of RA-ILD via inhibition of NFE2L1 PROTEIN in PBMC . However, their effectiveness in RA-ILD lacks further evidence from experiments and clinical trials. PRTN3 PROTEIN has been associated with RA progression and RA-ILD exacerbation and PADI1 PROTEIN (PAD2) is an enzyme that catalyzes protein citrullination. Increased levels of PAD2 were detected in lung homogenates in RA-ILD subjects. SDC1 PROTEIN was associated with internalization and degradation of TGF-β receptor in alveolar epithelial cells and was determined to exert antifibrotic effects on RA-ILD fibroblasts by regulating PAD2 (55, 127). Evidence from mouse model confirmed that PAD2 can be downregulated by SDC1 PROTEIN through PI3K/Akt signaling in lung fibroblasts. MMP19 PROTEIN (MMP9 PROTEIN) is known as a proteolytic enzyme that promotes the formation of synovial pannus and cartilage destruction in RA. Anti-MMP9 antibody andecaliximab can alleviate pulmonary fibrosis via TGF-β1-induced SMAD2 PROTEIN phosphorylation blockade in mouse model, exhibiting beneficial effects in patients with higher expression of type 1 IFNs in particular . Its safety and effectiveness in RA patients have been assessed through a Phase1b clinical trial . Meanwhile, target analysis and molecular docking to explain the effectiveness of triptolide and tripterine in CTD-ILD intervention suggest PI3K-Akt was identified as one of the main signaling pathways in molecular mechanism (130, 131). Findings demonstrated that it participates in regulation of fibrotic extracellular matrix (ECM) remodeling and therefore exhibits therapeutic value against pulmonary fibrosis . Paroxetine, a GRK2 PROTEIN inhibitor, was recently reported to suppress the activation of immune cells by inhibiting PI3K-AKT-mTOR pathway in patients with RA . Since GRK2 PROTEIN is also involved in TGF-β1-induced activation of lung fibroblasts, paroxetine is demonstrated to produce therapeutic effects in pulmonary fibrosis . Further clinical trials of patient cohort will be warranted. Previous opinions have considered PTEN PROTEIN as a critical ZMYND10 PROTEIN and also a crucial protector against alveolar epithelial injury and pulmonary fibrosis by controlling the PI3K/AKT signaling pathway (135, 136). Current findings on PTEN PROTEIN in RA pathogenesis indicated that its aberrant function contributes to fibroblast-like synoviocyte activation and bone destruction, which offers new opportunities to discover potential therapeutic targets . DIS3 PROTEIN (EXOSC4 PROTEIN), a component of the RNA exosome complex, was identified to be a potential autoantigen for RA-ILD patients. Recent findings suggested that levels of EXOSC4 antibodies were found to be significantly increased in sera of RA-ILD patients and it may contribute to the pulmonary fibrotic process via activation of Wnt/β-catenin signaling pathway . Knockdown of EXOSC4 PROTEIN can inhibit the activity of Wnt/β-catenin pathway and regulate the proliferation and differentiation of ATII cells, which provide valuable insights into potential mechanisms and therapy. Currently, it already serves as a therapeutic target in colorectal and ovarian cancers (139, 140). Meanwhile, since Wnt/β-catenin signaling pathway play a significant role in pulmonary fibrosis, other blockades of this pathway, including Calcaratarin D and Betulinic acid, were indicated to become novel antifibrotic agents for pulmonary fibrosis (141, 142). Osthole obtained from plants was estimated to restrain the function of RA-fibroblast-like synoviocytes through inhibition of NF-κB signaling, exerting a therapeutic effect in RA development. Meanwhile, by controlling accumulation and polarization of M2 macrophages in pulmonary interstitial fibrosis, it may also exhibit beneficial properties in RA patients with complications of ILD . Berberine has been reported with the potential to prevent RA and suppress inflammation through multiple signaling pathways, including Wnt/β-catenin, PI3K/Akt, and NF-κB . Its inhibitory effects to attenuate lung fibrosis through EMT and oxidative stress suppression as well as autophagy stimulation have also been confirmed by multiple relevant studies via animal models (145, 146). Further research and clinical trials on these pathway-related therapeutic targets exhibit great potential for improving the clinical outcomes of RA-ILD patients.7. Antifibrotic management Accumulating interest has been drawn to the application of antifibrotic medications against ILD progression, given the shared histological patterns and pathogenesis between IPF and RA-ILD. Pathways targeted by antifibrotic drugs, such as activation of TGF-β, WNT signal pathway and YAP/TAZ signal pathway, have become important strategies for fibrosis reduction . Novel therapeutic strategies focusing on the mechanism of fibrosis in RA-ILD, including nintedanib and pirfenidone, have recently been demonstrated to be promising in improving the outcomes of RA-ILD patients . Current viewpoints suggest the synergistic role of anti-fibrotic agents in treatment of RA-ILD combined with DMARDs. Evidence from the INBUILD trial has indicated that in groups treated with nintedanib, the rate of FVC decline was slower, both in patients with or without background therapy of DMARDs and/or glucocorticoids . Furthermore, it is suggested that compared with other fibrotic patterns, patients with UIP patterns shown by HRCT are more likely to benefit from nintedanib administration . However, patients under high-dose of immunosuppressive treatment were excluded from INBUILD probably due to safety concerns such as infection risks and hepatotoxicity. So far, data on high-dose combinations remain limited . Hepatotoxicity is one of the common adverse effects. An increase of ALT level over three times of the upper limit was reported in 15.1% patients treated with nintedanib . The potential effect of pirfenidone, another anti-fibrotic drug, was also being evaluated among RA-ILD patients. Similarly, studies recently have indicated their protective role against disease progression by slowing down FVC decline in RA-ILD patients . Pirfenidone can inhibit the transformation of fibroblasts into myofibroblasts and result in the reduction of critical cytokines involved in RA pathogenesis, including IL6 PROTEIN and TNFRSF1A PROTEIN (153, 154). While discontinuations have been reported in TRAIL1 due to adverse effects on gastrointestinal system (24.8%) . Other side effects include incidences of photosensitivity (12%) and rash (32%) in randomized trials . Further studies on antifibrotic drugs will be necessary to clarify their efficacy and safety in RA-ILD treatment. As for recent progress in cell therapy, rapid clinical improvement was observed in cases using CD19-targeting CAR-T cell therapy in treatment of CTD-ILD refractory to traditional immunosuppressive therapies . Last but not least, development of drug delivery systems based on RNAs, oligonucleotides, exosomes, and stem cells against pulmonary fibrosis is also worth noticing . For instance, one study developed mesenchymal stem cell-derived extracellular vesicles to treat pulmonary fibrosis by targeting senescent alveolar epithelial cells .8. Conclusions This review summarized the latest advances in RA-ILD, addressing the unsolved issues with great significance to inform future directions. Refined knowledge of the mechanistic interplay suggested a potential linkage of RA and ILD pathogenesis, involving multiple contributing factors. Meanwhile, progress in understanding the convergence of immunological pathways will contribute to a shared paradigm in patient management with considerations for both articular damage and the risk of lung fibrosis. Evidence of clinical evaluation from biomarkers is meaningful to allow for early identification from subclinical conditions, thereby predicting disease progression and guiding algorithms of optimal management. Currently, targeted therapies for RA treatment and their safety profile are being considered within clinical trials, while data regarding their use in RA-ILD as well as other efficacious options against pulmonary fibrosis are still limited. An urgent need remains for more studies on pathway-targeted therapeutic strategies to search for underlying opportunities that improve the overall outcomes of RA-ILD patients. Future research should prioritize the clinical validation of novel targeted therapies and exploration of combination therapies that concurrently address inflammation and fibrosis. By integrating these approaches, we can advance toward more effective, personalized treatment strategies that transform the management paradigm of RA-ILD.Acknowledgments We would like to thank Biorender.com for visual support.Funding Statement The author(s) declare financial support was received for the research and/or publication of this article. National Natural Science Foundation of China (No.82271825), Clinical Cohort Construction Program of Peking University Third Hospital No. BYSYDL2022017.Author contributions JX: Writing – original draft. CL: Writing – review & editing. JZ: Writing – review & editing. RM: Writing – review & editing.Conflict of interest The authors declare that the research was conducted in the absence of any commercial or financial relationships that could be construed as a potential conflict of interest.Generative AI statement The author(s) declare that no Generative AI was used in the creation of this manuscript. Any alternative text (alt text) provided alongside figures in this article has been generated by Frontiers with the support of artificial intelligence and reasonable efforts have been made to ensure accuracy, including review by the authors wherever possible. If you identify any issues, please contact us.Publisher’s note All claims expressed in this article are solely those of the authors and do not necessarily represent those of their affiliated organizations, or those of the publisher, the editors and the reviewers. Any product that may be evaluated in this article, or claim that may be made by its manufacturer, is not guaranteed or endorsed by the publisher.']"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_docs[0].page_content.split('References')[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "80cd4e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "#doc=WebBaseLoader('https://pmc.ncbi.nlm.nih.gov/articles/PMC11554381/').load()\n",
    "cleaned_docs=[]\n",
    "count=0\n",
    "for doc in all_docs:\n",
    "    if 'Reference' in doc.page_content and 'Abstract' in doc.page_content:\n",
    "        references=doc.page_content.split('.References')[:-1]\n",
    "        title=doc.metadata['title'].strip()\n",
    "        cleaning=re.sub(r\"\\s+\", \" \", doc.page_content)\n",
    "        cleaning=re.split(r\"Abstract\",cleaning)[1]\n",
    "        cleaning=re.split(r\"Acknowledgements\",cleaning)[0]\n",
    "        cleaning=re.sub(r'.Keywords*\\n?',' ',cleaning)\n",
    "        cleaning=re.sub('Open in a new tab','',cleaning)\n",
    "        cleaning=re.sub(r'[\\r\\n]+', r'\\n', cleaning)\n",
    "        cleaning=re.sub(r\"\\s*\\(?\\s*(?:Figure|Fig\\.?)\\s*[\\d.]+\\s*\\)?\\s*\",'',cleaning,flags=re.IGNORECASE).strip()\n",
    "        cleaning=re.sub(r'\\[[\\d\\s\\W]*?\\].','',cleaning,flags=re.IGNORECASE)\n",
    "        cleaning=re.sub(r\"\\(([^)]*?\\s*et\\s*al\\.\\s*[^)]*?)\\)\",'',cleaning)\n",
    "        cleaning=re.sub(r'\\(\\d+(?:,\\d+|-?\\d+)*\\)\\.','.',cleaning)\n",
    "        cleaning=re.sub(r'\\([\\w\\s]+\\d{4},?\\s*[\\d–-]*\\)','',cleaning)\n",
    "        cleaning=re.sub(r\"\\([rR]eviewed in\\)\",'',cleaning,flags=re.IGNORECASE)\n",
    "        cleaning=re.sub(r\"www\\.[^/]+/\",' ',cleaning)\n",
    "        cleaned_nlp=linker(nlp(cleaning))\n",
    "        new_text=cleaning\n",
    "        for i in reversed(cleaned_nlp.ents):\n",
    "            id=i._.kb_ents\n",
    "            if id:\n",
    "                name=linker.kb.cui_to_entity[id[0][0]].canonical_name\n",
    "                new_text = new_text[:i.start_char] + name + ' ' + i.label_ + new_text[i.end_char:]\n",
    "\n",
    "        doc.page_content=new_text\n",
    "        cleaned_docs.append(doc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54a28aff",
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_nlp=linker(nlp(cleaning))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "313e2871",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_text=cleaning\n",
    "for i in reversed(cleaned_nlp.ents):\n",
    "    id=i._.kb_ents\n",
    "    if id:\n",
    "        name=linker.kb.cui_to_entity[id[0][0]].canonical_name\n",
    "        new_text = new_text[:i.start_char] + name + ' ' + i.label_ + new_text[i.end_char:]\n",
    "\n",
    "doc[0].page_content=new_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fb30eec",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/bionik/AI_ML/FDA_Project/LLM/doc_list.pkl', 'wb') as f: # Open in write binary ('wb') mode\n",
    "    pickle.dump(cleaned_docs, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b587ee68",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/bionik/AI_ML/FDA_Project/LLM/doc_list.pkl', 'rb') as file:\n",
    "    cleaned_docs = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "44cb432c",
   "metadata": {},
   "outputs": [],
   "source": [
    "splitter = RecursiveCharacterTextSplitter(chunk_size=2000,chunk_overlap=500)\n",
    "chunks=splitter.split_documents(cleaned_docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "230263d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "embeddings_model = OllamaEmbeddings(model=\"qwen3-embedding:0.6b\",validate_model_on_init=True)\n",
    "vectorstore=FAISS.from_documents(chunks[:2],embeddings_model)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "43c7753f",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_chunks = [chunks[2:][i:i + 5] for i in range(0, len(chunks[2:]), 5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6490d989",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n",
      "362\n",
      "363\n",
      "364\n",
      "365\n",
      "366\n",
      "367\n",
      "368\n",
      "369\n",
      "370\n",
      "371\n",
      "372\n",
      "373\n",
      "374\n",
      "375\n",
      "376\n",
      "377\n",
      "378\n",
      "379\n",
      "380\n",
      "381\n",
      "382\n",
      "383\n",
      "384\n",
      "385\n",
      "386\n",
      "387\n",
      "388\n",
      "389\n",
      "390\n",
      "391\n",
      "392\n",
      "393\n",
      "394\n",
      "395\n",
      "396\n",
      "397\n",
      "398\n",
      "399\n",
      "400\n",
      "401\n",
      "402\n",
      "403\n",
      "404\n",
      "405\n",
      "406\n",
      "407\n",
      "408\n",
      "409\n",
      "410\n",
      "411\n",
      "412\n",
      "413\n",
      "414\n",
      "415\n",
      "416\n",
      "417\n",
      "418\n",
      "419\n",
      "420\n",
      "421\n",
      "422\n",
      "423\n",
      "424\n",
      "425\n",
      "426\n",
      "427\n",
      "428\n",
      "429\n",
      "430\n",
      "431\n",
      "432\n",
      "433\n",
      "434\n",
      "435\n",
      "436\n",
      "437\n",
      "438\n",
      "439\n",
      "440\n",
      "441\n",
      "442\n",
      "443\n",
      "444\n",
      "445\n",
      "446\n",
      "447\n",
      "448\n",
      "449\n",
      "450\n",
      "451\n",
      "452\n",
      "453\n",
      "454\n",
      "455\n",
      "456\n",
      "457\n",
      "458\n",
      "459\n",
      "460\n",
      "461\n",
      "462\n",
      "463\n",
      "464\n",
      "465\n",
      "466\n",
      "467\n",
      "468\n",
      "469\n",
      "470\n",
      "471\n",
      "472\n",
      "473\n",
      "474\n",
      "475\n",
      "476\n",
      "477\n",
      "478\n",
      "479\n",
      "480\n",
      "481\n",
      "482\n",
      "483\n",
      "484\n",
      "485\n",
      "486\n",
      "487\n",
      "488\n",
      "489\n",
      "490\n",
      "491\n"
     ]
    }
   ],
   "source": [
    "count=0\n",
    "for i in list_of_chunks:\n",
    "    try:\n",
    "        print(count)\n",
    "        count+=1\n",
    "        vectorstore.add_documents(i)\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2e51d31d",
   "metadata": {},
   "outputs": [],
   "source": [
    "retriever=vectorstore.as_retriever(search_type='similarity',search_kwargs={\"k\": 20, \"fetch_k\": 20})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e34b8051",
   "metadata": {},
   "outputs": [],
   "source": [
    "vectorstore.save_local(\"/home/bionik/AI_ML/FDA_Project/LLM/faiss_index\")"
   ]
  },
  {
   "cell_type": "raw",
   "id": "252be7f8",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "vectorstore = FAISS.load_local(\n",
    "    \"/home/bionik/AI_ML/FDA_Project/Data/LLM/faiss_index\",\n",
    "    embeddings_model,\n",
    "    allow_dangerous_deserialization=True \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ddf5ad38",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Triplet(BaseModel):\n",
    "    subject: str = Field(..., description=\"ONE protein or ONE protein complex\")\n",
    "    predicate: str = Field(..., description=\"Relationship between protein entities\")\n",
    "    object: str = Field(..., description=\"ONE protein or ONE protein complex\")\n",
    "\n",
    "class TripletList(BaseModel):\n",
    "    triplets: List[Triplet]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "0b7e3f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"\"\"\n",
    "You are an expert information extraction assistant. \n",
    "Your task is to extract knowledge triples in the format of (subject, predicate, object) from the provided text. A subject and object are protein entities, and the predicate describes the relationship between them.\n",
    "\n",
    "**Instructions:**\n",
    "1.  Identify all relevant subject-predicate-object triples that represent facts stated explicitly in the text.\n",
    "2.  Focus on extracting facts accurately and completely.\n",
    "3.  Do not hallucinate or add information that is not present in the source text.\n",
    "4.  If a sentence contains multiple facts, extract each fact as a separate triple.\n",
    "5.  Predicate values should be only from ['Activates','Inhibits','Reacts','Form Complex']\n",
    "6.  Subject and object should be only proteins.\n",
    "\n",
    "Normalize synonyms:\n",
    "binds/interacts/phosphorylate → Reacts\n",
    "forms a complex → Form Complex\n",
    "suppresses/destabilizes/activates degradation → Inhibits\n",
    "induces/promotes/activates/stabilizes/inhibits degradation → Activates\n",
    "\n",
    "Retrieve triplets from this text below:\n",
    "{context}\n",
    "\n",
    "\"\"\"\n",
    "prompt = ChatPromptTemplate.from_template(template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15b4a834",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm=ChatOllama(model='qwen3:4b-instruct-2507-q4_K_M',temperature=0.1,validate_model_on_init=True,num_ctx=12000) ## this is better\n",
    "#llm=ChatOllama(model='qwen3:4b-instruct-2507-q4_K_M',temperature=0.0,validate_model_on_init=True,num_ctx=12000).with_structured_output(TripletList) ## this is better\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e89e2060",
   "metadata": {},
   "outputs": [],
   "source": [
    "rag_chain = (RunnableParallel(context=retriever,question=RunnablePassthrough())\n",
    "    | prompt\n",
    "    | llm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "20a23e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = rag_chain.invoke(\"What are the CDC37 proteins's key characteristics/interactions?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83d63180",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df262475",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01e0e289",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddb3c86f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7b0213d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "system_template = \"\"\"\n",
    "Use the following pieces of retrieved context to answer the question of research article of biology. \n",
    "If you don't know the answer, genuinely say that you don't know.\n",
    "Use only the information which was provided in this query\n",
    "Keep the answer detailed and precise.\n",
    "\n",
    "The question is:\n",
    "{question}\n",
    "\n",
    "Use this context to answer the question:\n",
    "{context}\n",
    "\"\"\"\n",
    "\n",
    "prompt = ChatPromptTemplate.from_messages([SystemMessagePromptTemplate.from_template(system_template),\n",
    "                                           HumanMessagePromptTemplate.from_template('{question}')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7d2c14e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "rag_chain = (\n",
    "    RunnableParallel({\"context\": retriever, \"question\": RunnablePassthrough()})\n",
    "    | prompt\n",
    "    | llm\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f6e7aa",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[43], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m ans\u001b[38;5;241m=\u001b[39m\u001b[43mllm\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mexplain details EMT and which pathways are involved in this process?\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py:398\u001b[0m, in \u001b[0;36mBaseChatModel.invoke\u001b[0;34m(self, input, config, stop, **kwargs)\u001b[0m\n\u001b[1;32m    384\u001b[0m \u001b[38;5;129m@override\u001b[39m\n\u001b[1;32m    385\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21minvoke\u001b[39m(\n\u001b[1;32m    386\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    391\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m    392\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m AIMessage:\n\u001b[1;32m    393\u001b[0m     config \u001b[38;5;241m=\u001b[39m ensure_config(config)\n\u001b[1;32m    394\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m cast(\n\u001b[1;32m    395\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAIMessage\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    396\u001b[0m         cast(\n\u001b[1;32m    397\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mChatGeneration\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m--> 398\u001b[0m             \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate_prompt\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    399\u001b[0m \u001b[43m                \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_convert_input\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    400\u001b[0m \u001b[43m                \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    401\u001b[0m \u001b[43m                \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcallbacks\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    402\u001b[0m \u001b[43m                \u001b[49m\u001b[43mtags\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtags\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    403\u001b[0m \u001b[43m                \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmetadata\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    404\u001b[0m \u001b[43m                \u001b[49m\u001b[43mrun_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrun_name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    405\u001b[0m \u001b[43m                \u001b[49m\u001b[43mrun_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpop\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrun_id\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    406\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    407\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mgenerations[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m0\u001b[39m],\n\u001b[1;32m    408\u001b[0m         )\u001b[38;5;241m.\u001b[39mmessage,\n\u001b[1;32m    409\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py:1117\u001b[0m, in \u001b[0;36mBaseChatModel.generate_prompt\u001b[0;34m(self, prompts, stop, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m   1108\u001b[0m \u001b[38;5;129m@override\u001b[39m\n\u001b[1;32m   1109\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mgenerate_prompt\u001b[39m(\n\u001b[1;32m   1110\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1114\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m   1115\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m LLMResult:\n\u001b[1;32m   1116\u001b[0m     prompt_messages \u001b[38;5;241m=\u001b[39m [p\u001b[38;5;241m.\u001b[39mto_messages() \u001b[38;5;28;01mfor\u001b[39;00m p \u001b[38;5;129;01min\u001b[39;00m prompts]\n\u001b[0;32m-> 1117\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprompt_messages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py:927\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[0;34m(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)\u001b[0m\n\u001b[1;32m    924\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(input_messages):\n\u001b[1;32m    925\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    926\u001b[0m         results\u001b[38;5;241m.\u001b[39mappend(\n\u001b[0;32m--> 927\u001b[0m             \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_generate_with_cache\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    928\u001b[0m \u001b[43m                \u001b[49m\u001b[43mm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    929\u001b[0m \u001b[43m                \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    930\u001b[0m \u001b[43m                \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_managers\u001b[49m\u001b[43m[\u001b[49m\u001b[43mi\u001b[49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mrun_managers\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    931\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    932\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    933\u001b[0m         )\n\u001b[1;32m    934\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    935\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m run_managers:\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py:1221\u001b[0m, in \u001b[0;36mBaseChatModel._generate_with_cache\u001b[0;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m   1219\u001b[0m     result \u001b[38;5;241m=\u001b[39m generate_from_stream(\u001b[38;5;28miter\u001b[39m(chunks))\n\u001b[1;32m   1220\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m inspect\u001b[38;5;241m.\u001b[39msignature(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate)\u001b[38;5;241m.\u001b[39mparameters\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_manager\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m-> 1221\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_generate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1222\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmessages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_manager\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\n\u001b[1;32m   1223\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1224\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1225\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate(messages, stop\u001b[38;5;241m=\u001b[39mstop, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/langchain_ollama/chat_models.py:1030\u001b[0m, in \u001b[0;36mChatOllama._generate\u001b[0;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m   1023\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_generate\u001b[39m(\n\u001b[1;32m   1024\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1025\u001b[0m     messages: \u001b[38;5;28mlist\u001b[39m[BaseMessage],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1028\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m   1029\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ChatResult:\n\u001b[0;32m-> 1030\u001b[0m     final_chunk \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_chat_stream_with_aggregation\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1031\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmessages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\n\u001b[1;32m   1032\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1033\u001b[0m     generation_info \u001b[38;5;241m=\u001b[39m final_chunk\u001b[38;5;241m.\u001b[39mgeneration_info\n\u001b[1;32m   1034\u001b[0m     chat_generation \u001b[38;5;241m=\u001b[39m ChatGeneration(\n\u001b[1;32m   1035\u001b[0m         message\u001b[38;5;241m=\u001b[39mAIMessage(\n\u001b[1;32m   1036\u001b[0m             content\u001b[38;5;241m=\u001b[39mfinal_chunk\u001b[38;5;241m.\u001b[39mtext,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1043\u001b[0m         generation_info\u001b[38;5;241m=\u001b[39mgeneration_info,\n\u001b[1;32m   1044\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/langchain_ollama/chat_models.py:965\u001b[0m, in \u001b[0;36mChatOllama._chat_stream_with_aggregation\u001b[0;34m(self, messages, stop, run_manager, verbose, **kwargs)\u001b[0m\n\u001b[1;32m    956\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_chat_stream_with_aggregation\u001b[39m(\n\u001b[1;32m    957\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    958\u001b[0m     messages: \u001b[38;5;28mlist\u001b[39m[BaseMessage],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    962\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m    963\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ChatGenerationChunk:\n\u001b[1;32m    964\u001b[0m     final_chunk \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 965\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m chunk \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_iterate_over_stream(messages, stop, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    966\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m final_chunk \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    967\u001b[0m             final_chunk \u001b[38;5;241m=\u001b[39m chunk\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/langchain_ollama/chat_models.py:1054\u001b[0m, in \u001b[0;36mChatOllama._iterate_over_stream\u001b[0;34m(self, messages, stop, **kwargs)\u001b[0m\n\u001b[1;32m   1047\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_iterate_over_stream\u001b[39m(\n\u001b[1;32m   1048\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1049\u001b[0m     messages: \u001b[38;5;28mlist\u001b[39m[BaseMessage],\n\u001b[1;32m   1050\u001b[0m     stop: \u001b[38;5;28mlist\u001b[39m[\u001b[38;5;28mstr\u001b[39m] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   1051\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m   1052\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Iterator[ChatGenerationChunk]:\n\u001b[1;32m   1053\u001b[0m     reasoning \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreasoning\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreasoning)\n\u001b[0;32m-> 1054\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m stream_resp \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_create_chat_stream(messages, stop, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m   1055\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(stream_resp, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m   1056\u001b[0m             content \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m   1057\u001b[0m                 stream_resp[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessage\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1058\u001b[0m                 \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessage\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m stream_resp \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m stream_resp[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessage\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1059\u001b[0m                 \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1060\u001b[0m             )\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/langchain_ollama/chat_models.py:952\u001b[0m, in \u001b[0;36mChatOllama._create_chat_stream\u001b[0;34m(self, messages, stop, **kwargs)\u001b[0m\n\u001b[1;32m    950\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chat_params[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m    951\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_client:\n\u001b[0;32m--> 952\u001b[0m         \u001b[38;5;28;01myield from\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_client\u001b[38;5;241m.\u001b[39mchat(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mchat_params)\n\u001b[1;32m    953\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_client:\n\u001b[1;32m    954\u001b[0m     \u001b[38;5;28;01myield\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_client\u001b[38;5;241m.\u001b[39mchat(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mchat_params)\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/ollama/_client.py:181\u001b[0m, in \u001b[0;36mClient._request.<locals>.inner\u001b[0;34m()\u001b[0m\n\u001b[1;32m    178\u001b[0m   e\u001b[38;5;241m.\u001b[39mresponse\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m    179\u001b[0m   \u001b[38;5;28;01mraise\u001b[39;00m ResponseError(e\u001b[38;5;241m.\u001b[39mresponse\u001b[38;5;241m.\u001b[39mtext, e\u001b[38;5;241m.\u001b[39mresponse\u001b[38;5;241m.\u001b[39mstatus_code) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 181\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m r\u001b[38;5;241m.\u001b[39miter_lines():\n\u001b[1;32m    182\u001b[0m   part \u001b[38;5;241m=\u001b[39m json\u001b[38;5;241m.\u001b[39mloads(line)\n\u001b[1;32m    183\u001b[0m   \u001b[38;5;28;01mif\u001b[39;00m err \u001b[38;5;241m:=\u001b[39m part\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124merror\u001b[39m\u001b[38;5;124m'\u001b[39m):\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpx/_models.py:929\u001b[0m, in \u001b[0;36mResponse.iter_lines\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    927\u001b[0m decoder \u001b[38;5;241m=\u001b[39m LineDecoder()\n\u001b[1;32m    928\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m request_context(request\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_request):\n\u001b[0;32m--> 929\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m text \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39miter_text():\n\u001b[1;32m    930\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m decoder\u001b[38;5;241m.\u001b[39mdecode(text):\n\u001b[1;32m    931\u001b[0m             \u001b[38;5;28;01myield\u001b[39;00m line\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpx/_models.py:916\u001b[0m, in \u001b[0;36mResponse.iter_text\u001b[0;34m(self, chunk_size)\u001b[0m\n\u001b[1;32m    914\u001b[0m chunker \u001b[38;5;241m=\u001b[39m TextChunker(chunk_size\u001b[38;5;241m=\u001b[39mchunk_size)\n\u001b[1;32m    915\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m request_context(request\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_request):\n\u001b[0;32m--> 916\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m byte_content \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39miter_bytes():\n\u001b[1;32m    917\u001b[0m         text_content \u001b[38;5;241m=\u001b[39m decoder\u001b[38;5;241m.\u001b[39mdecode(byte_content)\n\u001b[1;32m    918\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m chunk \u001b[38;5;129;01min\u001b[39;00m chunker\u001b[38;5;241m.\u001b[39mdecode(text_content):\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpx/_models.py:897\u001b[0m, in \u001b[0;36mResponse.iter_bytes\u001b[0;34m(self, chunk_size)\u001b[0m\n\u001b[1;32m    895\u001b[0m chunker \u001b[38;5;241m=\u001b[39m ByteChunker(chunk_size\u001b[38;5;241m=\u001b[39mchunk_size)\n\u001b[1;32m    896\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m request_context(request\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_request):\n\u001b[0;32m--> 897\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m raw_bytes \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39miter_raw():\n\u001b[1;32m    898\u001b[0m         decoded \u001b[38;5;241m=\u001b[39m decoder\u001b[38;5;241m.\u001b[39mdecode(raw_bytes)\n\u001b[1;32m    899\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m chunk \u001b[38;5;129;01min\u001b[39;00m chunker\u001b[38;5;241m.\u001b[39mdecode(decoded):\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpx/_models.py:951\u001b[0m, in \u001b[0;36mResponse.iter_raw\u001b[0;34m(self, chunk_size)\u001b[0m\n\u001b[1;32m    948\u001b[0m chunker \u001b[38;5;241m=\u001b[39m ByteChunker(chunk_size\u001b[38;5;241m=\u001b[39mchunk_size)\n\u001b[1;32m    950\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m request_context(request\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_request):\n\u001b[0;32m--> 951\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m raw_stream_bytes \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstream:\n\u001b[1;32m    952\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_num_bytes_downloaded \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(raw_stream_bytes)\n\u001b[1;32m    953\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m chunk \u001b[38;5;129;01min\u001b[39;00m chunker\u001b[38;5;241m.\u001b[39mdecode(raw_stream_bytes):\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpx/_client.py:153\u001b[0m, in \u001b[0;36mBoundSyncStream.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m__iter__\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m typing\u001b[38;5;241m.\u001b[39mIterator[\u001b[38;5;28mbytes\u001b[39m]:\n\u001b[0;32m--> 153\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m chunk \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_stream:\n\u001b[1;32m    154\u001b[0m         \u001b[38;5;28;01myield\u001b[39;00m chunk\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpx/_transports/default.py:127\u001b[0m, in \u001b[0;36mResponseStream.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    125\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m__iter__\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m typing\u001b[38;5;241m.\u001b[39mIterator[\u001b[38;5;28mbytes\u001b[39m]:\n\u001b[1;32m    126\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m map_httpcore_exceptions():\n\u001b[0;32m--> 127\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m part \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_httpcore_stream:\n\u001b[1;32m    128\u001b[0m             \u001b[38;5;28;01myield\u001b[39;00m part\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py:407\u001b[0m, in \u001b[0;36mPoolByteStream.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    405\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    406\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m--> 407\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py:403\u001b[0m, in \u001b[0;36mPoolByteStream.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    401\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m__iter__\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m typing\u001b[38;5;241m.\u001b[39mIterator[\u001b[38;5;28mbytes\u001b[39m]:\n\u001b[1;32m    402\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 403\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m part \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_stream:\n\u001b[1;32m    404\u001b[0m             \u001b[38;5;28;01myield\u001b[39;00m part\n\u001b[1;32m    405\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpcore/_sync/http11.py:342\u001b[0m, in \u001b[0;36mHTTP11ConnectionByteStream.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    340\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m ShieldCancellation():\n\u001b[1;32m    341\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m--> 342\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m exc\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpcore/_sync/http11.py:334\u001b[0m, in \u001b[0;36mHTTP11ConnectionByteStream.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    332\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    333\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Trace(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreceive_response_body\u001b[39m\u001b[38;5;124m\"\u001b[39m, logger, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_request, kwargs):\n\u001b[0;32m--> 334\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m chunk \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_connection\u001b[38;5;241m.\u001b[39m_receive_response_body(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    335\u001b[0m             \u001b[38;5;28;01myield\u001b[39;00m chunk\n\u001b[1;32m    336\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    337\u001b[0m     \u001b[38;5;66;03m# If we get an exception while streaming the response,\u001b[39;00m\n\u001b[1;32m    338\u001b[0m     \u001b[38;5;66;03m# we want to close the response (and possibly the connection)\u001b[39;00m\n\u001b[1;32m    339\u001b[0m     \u001b[38;5;66;03m# before raising that exception.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpcore/_sync/http11.py:203\u001b[0m, in \u001b[0;36mHTTP11Connection._receive_response_body\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    200\u001b[0m timeout \u001b[38;5;241m=\u001b[39m timeouts\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mread\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m    202\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 203\u001b[0m     event \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_receive_event\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    204\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(event, h11\u001b[38;5;241m.\u001b[39mData):\n\u001b[1;32m    205\u001b[0m         \u001b[38;5;28;01myield\u001b[39;00m \u001b[38;5;28mbytes\u001b[39m(event\u001b[38;5;241m.\u001b[39mdata)\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpcore/_sync/http11.py:217\u001b[0m, in \u001b[0;36mHTTP11Connection._receive_event\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    214\u001b[0m     event \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_h11_state\u001b[38;5;241m.\u001b[39mnext_event()\n\u001b[1;32m    216\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m event \u001b[38;5;129;01mis\u001b[39;00m h11\u001b[38;5;241m.\u001b[39mNEED_DATA:\n\u001b[0;32m--> 217\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_network_stream\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    218\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mREAD_NUM_BYTES\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    219\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    221\u001b[0m     \u001b[38;5;66;03m# If we feed this case through h11 we'll raise an exception like:\u001b[39;00m\n\u001b[1;32m    222\u001b[0m     \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m    223\u001b[0m     \u001b[38;5;66;03m#     httpcore.RemoteProtocolError: can't handle event type\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    227\u001b[0m     \u001b[38;5;66;03m# perspective. Instead we handle this case distinctly and treat\u001b[39;00m\n\u001b[1;32m    228\u001b[0m     \u001b[38;5;66;03m# it as a ConnectError.\u001b[39;00m\n\u001b[1;32m    229\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m data \u001b[38;5;241m==\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_h11_state\u001b[38;5;241m.\u001b[39mtheir_state \u001b[38;5;241m==\u001b[39m h11\u001b[38;5;241m.\u001b[39mSEND_RESPONSE:\n",
      "File \u001b[0;32m~/miniconda3/envs/LLM/lib/python3.10/site-packages/httpcore/_backends/sync.py:128\u001b[0m, in \u001b[0;36mSyncStream.read\u001b[0;34m(self, max_bytes, timeout)\u001b[0m\n\u001b[1;32m    126\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_exceptions(exc_map):\n\u001b[1;32m    127\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sock\u001b[38;5;241m.\u001b[39msettimeout(timeout)\n\u001b[0;32m--> 128\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmax_bytes\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "ans=llm.invoke('explain details EMT and which pathways are involved in this process?')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0c025e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "ans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19b04f11",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_data = 'explain details EMT and which pathways are involved in this process?'\n",
    "response = rag_chain.invoke(input_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5ae5629",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'input_tokens': 21, 'output_tokens': 1442, 'total_tokens': 1463}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3eba0939",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(id='977c2c03-2d89-49a0-b10b-ddf2fda5c124', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/9822656/', 'title': '\\n            Acute Myeloid Leukemia Following Myeloproliferative Neoplasms: A Review of What We Know, What We Do Not Know, and Emerging Treatment Strategies - PMC\\n        ', 'description': 'Acute myeloid leukemia (AML) arising from myeloproliferative neoplasms (MPNs) represents a small subtype of secondary AML (sAML). This entity is well known to be associated with poor responses to available treatment options and dismal outcomes. To ...', 'language': 'en'}, page_content='collection Name your collection * Choose a collection Unable to load your collection due to an error Please try again Add Cancel Follow NCBI NCBI on X (formerly known as Twitter) NCBI on Facebook NCBI on LinkedIn NCBI on GitHub NCBI RSS feed Connect with NLM NLM on X (formerly known as Twitter) NLM on Facebook NLM on YouTube National Library of Medicine 8600 Rockville Pike Bethesda, MD 20894 Web Policies FOIA HHS Vulnerability Disclosure Help Accessibility Careers NLM NIH HHS USA.gov Back to Top'),\n",
       " Document(id='6d0ce718-07a0-4c27-9db1-2ef12331b613', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/4867197/', 'title': '\\n            The role of the Janus kinase family/signal transducer and activator of transcription signaling pathway in fibrotic renal disease - PMC\\n        ', 'description': 'Over the past several years, a number of cytokines and growth factors including transforming growth factor β1, tumor necrosis factor α, and angiotensin II have been shown to play a crucial role in renal fibrosis. The Janus kinase family (JAK) and ...', 'language': 'en'}, page_content='derived from renal TECs during pathologic states via the process of EMT  Using S100A4 PROTEIN (S100A4 PROTEIN), a marker that is normally expressed by fibroblasts but not in epithelial cells, Strutz et al. showed that TECs could express S100A4 PROTEIN in a mouse model of antitubular basement membrane disease. This was the first study to demonstrate the capability of epithelial cells to transform into fibroblast-like cells. A subsequent study by Iwano et al. revealed that a substantial number of interstitial fibroblasts originate from tubular epithelium (EMT) during renal fibrosis. Of the Fsp1-positive fibroblast pool in the fibrotic kidney, 36% coexpressed LacZ that had been tagged to renal proximal tubules, providing proof that fibroblasts can form locally by EMT during the pathologic stress of tissue fibrosis  TGF-β1 is a major regulator of tubulointerstitial fibrosis  TGF-β1 can independently initiate and complete the entire course of EMT in vitro. In addition, TGF-β1 stimulates ECM synthesis while simultaneously inhibiting MMP13 PROTEIN and degradative matrix metalloproteinases in vivo  TNF-α has also been implicated in fibrotic renal injury, stimulating ECM accumulation, inhibiting ECM degradation, and upregulating a number of cytokines and NFE2L1 PROTEIN involved in tubulointerstitial fibrosis  Recently, IL18 PROTEIN has been identified as an important mediator of obstruction-induced renal fibrosis, ECM deposition, and EMT, independent of TGF-β1 or TNF-α activity  Despite the clear role of IL9 PROTEIN signaling in obstructive renal injury, few studies have investigated the potential contribution of JAK/STAT signaling to obstructive nephropathy  Kuratsune et al. have examined STAT3 PROTEIN activation in a rat model of unilateral ureteral obstruction. The authors demonstrated STAT3 PROTEIN activation in renal TECs and myofibroblasts in response to 3 or 7 days of obstruction, with peak STAT3 PROTEIN expression occurring 7 days after the onset of obstruction'),\n",
       " Document(id='0d965ba2-5dc5-47b0-be03-16ae39e3a887', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/12741264/', 'title': '\\n            Comprehensive Review of Sepsis-Related Skeletal Muscle Atrophy: Mechanisms, Diagnosis and Therapeutic Strategies - PMC\\n        ', 'description': 'This comprehensive review aims to synthesize current evidence on sepsis-related skeletal muscle atrophy, with a focus on underlying mechanisms, diagnostic approaches, and therapeutic interventions. A systematic literature search was conducted in ...', 'language': 'en'}, page_content='to Collections Create a new collection Add to an existing collection Name your collection * Choose a collection Unable to load your collection due to an error Please try again Add Cancel Follow NCBI NCBI on X (formerly known as Twitter) NCBI on Facebook NCBI on LinkedIn NCBI on GitHub NCBI RSS feed Connect with NLM NLM on X (formerly known as Twitter) NLM on Facebook NLM on YouTube National Library of Medicine 8600 Rockville Pike Bethesda, MD 20894 Web Policies FOIA HHS Vulnerability Disclosure Help Accessibility Careers NLM NIH HHS USA.gov Back to Top'),\n",
       " Document(id='5f9ad96c-9fe2-4329-afe1-9562d1339863', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/11433914/', 'title': '\\n            Spastin and alsin protein interactome analyses begin to reveal key canonical pathways and suggest novel druggable targets - PMC\\n        ', 'description': 'Developing effective and long-term treatment strategies for rare and complex neurodegenerative diseases is challenging. One of the major roadblocks is the extensive heterogeneity among patients. This hinders understanding the underlying ...', 'language': 'en'}, page_content='Molecular interactions Endosomal sorting complex required for transport (ESCRT) Full length HD MIT AAA None - Mitotic metaphase and anaphase Full lengthHDMIT Bortezomib/paclitaxelDocetaxel/irinotecanL-asparaginase/daunorubicin/prednisone/vincristineAfatinib/paclitaxelVinorelbinePaclitaxel/pembrolizumabEpothilone BColchicineVincristineDocetaxel/5-fluorouracil/oxaliplatin/trastuzumab 6555555555 Late endosomal microautophagy Full lengthAAAMIT LumacaftorIvacaftor/lumacaftorTributyrinCFTR modulatorBortezomib/vorinostatVorinostatJBI-802Ivacaftor/lumacaftorKA2507Ivacaftor/tezacaftor 4444444444 Iron uptake and transport Full length HD SLC3A2 PROTEIN A27.15CX-2029Oxaliplatin-encapsulated transferrin-conjugated-N-Glutaryl phosphatidylethanolamine liposomeTLD-1433Ferric carboxymaltoseNKP-1339PlatinumMORAb-066Delta-aminolevulinic acid 8888777777 Mechanisms of viral exit from host cells Full lengthAAAHD WJ01024Ingenol mebutateRo31-8425FelezonexorBortezomib/dexamethasone/selinexorBryostatin 1Myristoylated protein kinase C peptide inhibitorSotrastaurinAprinocarsenProtein kinase C inhibitor 1111111111 RHO RHOU PROTEIN cycle AAA GSK1278893ASM1-71CALAA-01CX-2029Monoclonal antibody A27.15Oxaliplatin-encapsulatedtransferrin-conjugated-N-Glutaryl phosphatidylethanolamine liposomeCytarabine/dasatinib/dexamethasone/methotrexateDasatinibDasatinib/dexamethasoneBlinatumomab/dasatinib 34272626262621212121 Macroautophagy AAA WZ3105SBI-0206965MRT68921DCC-3116PIK-IIIVPS34-IN1VSP34 inhibitor 1SM1-71SAR405mTOR inhibitor 10101010777776 Metabolism of MIR124-1HG RNA HD RisdiplamNusinersenDusquetideJNJ-64619178PF-06939999MRTX1719PRMT5 inhibitorTNG908SH3765GSK3326595 6611111111 RIPK1 PROTEIN mediated regulated necrosis MIT GSK872GSK840GSK843GSK3145095EclitasertibGSK963GSK2982772APO010MapatumumabLuminespib 27272724242424333 ISG15 PROTEIN antiviral mechanism MIT INCB-16562HE3286SolcitinibTofacitinibFilgotinibJaktinibBrepocitinibKO-947SAR-20347INCB052793 2222222222 Clathrin mediated endocytosis MTBD'),\n",
       " Document(id='926a9452-c1c4-4aa0-8ba0-ec2984af75fb', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/11433914/', 'title': '\\n            Spastin and alsin protein interactome analyses begin to reveal key canonical pathways and suggest novel druggable targets - PMC\\n        ', 'description': 'Developing effective and long-term treatment strategies for rare and complex neurodegenerative diseases is challenging. One of the major roadblocks is the extensive heterogeneity among patients. This hinders understanding the underlying ...', 'language': 'en'}, page_content='PROTEIN pathway define a molecular subset of rapidly progressing lung adenocarcinoma. J Thorac Oncol. 2019;14:1924–1934. doi: 10.1016/j.jtho.2019.07.003. [DOI] [PubMed] [Google Scholar] Gozutok O, Helmold BR, Ozdinler PH. Mutations and protein interaction landscape reveal key cellular events perturbed in upper motor neurons with HSP and PLS. Brain Sci. 2021;11:578. doi: 10.3390/brainsci11050578. [DOI] [PMC free article] [PubMed] [Google Scholar] Hakes L, Lovell SC, Oliver SG, Robertson DL. Specificity in protein interactions and its relationship with sequence diversity and coevolution. Proc Natl Acad Sci U S A. 2007;104:7999–8004. doi: 10.1073/pnas.0609962104. [DOI] [PMC free article] [PubMed] [Google Scholar] Helal M, Mazaheri N, Shalbafan B, Malamiri RA, Dilaver N, Buchert R, Mohammadiasl J, Golchin N, Sedaghat A, Mehrjardi MYV, Haack TB, Riess O, Chung WK, Galehdari H, Shariati G, Maroofian R. Clinical presentation and natural history of infantile-onset ascending spastic paralysis from three families with an ALS2 founder variant. Neurol Sci. 2018;39:1917–1925. doi: 10.1007/s10072-018-3526-8. [DOI] [PubMed] [Google Scholar] Helmold BR, Pauss KE, Ozdinler PH. TARDBP PROTEIN protein interactome informs about perturbed canonical pathways and may help develop personalized medicine approaches for patients with TARDBP PROTEIN pathology. Drug Discov Today. 2023;28:103769. doi: 10.1016/j.drudis.2023.103769. [DOI] [PMC free article] [PubMed] [Google Scholar] Henne WM, Buchkovich NJ, Emr SD. The ESCRT pathway. Dev Cell. 2011;21:77–91. doi: 10.1016/j.devcel.2011.05.015. [DOI] [PubMed] [Google Scholar] Honorato RV, Koukos PI, Jimenez-Garcia B, Tsaregorodtsev A, Verlato M, Giachetti A, Rosato A, Bonvin A. Structural biology in the clouds: the WeNMR-EOSC ecosystem. Front Mol Biosci. 2021;8:729513. doi: 10.3389/fmolb.2021.729513. [DOI] [PMC free article] [PubMed] [Google Scholar] Hsu F, Spannl S, Ferguson C, Hyman AA, Parton RG, RAB5A PROTEIN and ALS2 PROTEIN regulate'),\n",
       " Document(id='de4b19e9-cbb9-4d03-b69a-189db478a619', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/11433914/', 'title': '\\n            Spastin and alsin protein interactome analyses begin to reveal key canonical pathways and suggest novel druggable targets - PMC\\n        ', 'description': 'Developing effective and long-term treatment strategies for rare and complex neurodegenerative diseases is challenging. One of the major roadblocks is the extensive heterogeneity among patients. This hinders understanding the underlying ...', 'language': 'en'}, page_content='Mol Basis Dis. 20171863:1666–1677. doi: 10.1016/j.bbadis.2017.04.007. [DOI] [PubMed] [Google Scholar] Plewczynski D, Ginalski K. The interactome: predicting the protein-protein interactions in cells. Cell Mol Biol Lett. 2009;14:1–22. doi: 10.2478/s11658-008-0024-7. [DOI] [PMC free article] [PubMed] [Google Scholar] Prichard KL, O’Brien NS, Murcia SR, Baker JR, McCluskey A. Role of clathrin and dynamin in clathrin mediated endocytosis/synaptic vesicle recycling and implications in neurological diseases. Front Cell Neurosci. 2021;15:754110. doi: 10.3389/fncel.2021.754110. [DOI] [PMC free article] [PubMed] [Google Scholar] Rai SN, Dilnashin H, Birla H, Singh SS, Zahra W, Rathore AS, Singh BK, Singh SP. The role of PI3K/Akt and EPHB2 PROTEIN in neurodegenerative disorders. Neurotox Res. 2019;35:775–795. doi: 10.1007/s12640-019-0003-y. [DOI] [PubMed] [Google Scholar] Ralli M, Lambiase A, Artico M, de Vincentiis M, Greco A. Amyotrophic lateral sclerosis: autoimmune pathogenic mechanisms, clinical features, and therapeutic perspectives. Isr Med Assoc J. 2019;21:438–443. [PubMed] [Google Scholar] Rand JM, Pisithkul T, Clark RL, Thiede JM, Mehrer CR, Agnew DE, Campbell CE, Markley AL, Price MN, Ray J, Wetmore KM, Suh Y, Arkin AP, Deutschbauer AM, Amador-Noguez D, Pfleger BF. A metabolic pathway for catabolizing levulinic acid in bacteria. Nat Microbiol. 2017;2:1624–1634. doi: 10.1038/s41564-017-0028-z. [DOI] [PMC free article] [PubMed] [Google Scholar] Rao S, Du G, Hafner M, Subramanian K, Sorger PK, Gray NS. A multitargeted probe-based strategy to identify signaling vulnerabilities in cancers. J Biol Chem. 2019;294:8664–8673. doi: 10.1074/jbc.RA118.006805. [DOI] [PMC free article] [PubMed] [Google Scholar] Rao VS, Srinivas K, Sujini GN, Kumar GN. Protein-protein interaction detection: methods and analysis. Int J Proteomics. 2014;2014:147648. doi: 10.1155/2014/147648. [DOI] [PMC free article] [PubMed] [Google Scholar] Roll-Mecak A, McNally FJ. Microtubule-severing enzymes.'),\n",
       " Document(id='04cf32ac-f075-4803-9b3c-afacc77f0792', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/12729966/', 'title': '\\n            Integrated Transcriptomic and Proteomic Analysis Associated with Knockdown and Overexpression Studies Revealed ECHDC1 as a Regulator of Intramuscular Fat Deposition in Cattle - PMC\\n        ', 'description': 'Intramuscular fat (IMF) content is a critical determinant of beef flavor, tenderness, and juiciness. It is regulated by many factors, including nutrition, non-coding RNA, and genetic factors, among which, genetic factors play a key role. Therefore, ...', 'language': 'en'}, page_content='ACTIONS View on publisher PDF DNA (2.2 MB) Cite Collections Permalink PERMALINK Copy RESOURCES Similar articles Cited by other articles Links to NCBI Databases Cite Copy Download .nbib .nbib Format: AMA APA MLA NLM Add to Collections Create a new collection Add to an existing collection Name your collection * Choose a collection Unable to load your collection due to an error Please try again Add Cancel Follow NCBI NCBI on X (formerly known as Twitter) NCBI on Facebook NCBI on LinkedIn NCBI on GitHub NCBI RSS feed Connect with NLM NLM on X (formerly known as Twitter) NLM on Facebook NLM on YouTube National Library of Medicine 8600 Rockville Pike Bethesda, MD 20894 Web Policies FOIA HHS Vulnerability Disclosure Help Accessibility Careers NLM NIH HHS USA.gov Back to Top'),\n",
       " Document(id='a5695710-450d-4953-a999-afc18d0fbc24', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/9345973/', 'title': '\\n            Bioinformatics analysis reveals potential biomarkers associated with the occurrence of intracranial aneurysms - PMC\\n        ', 'description': 'To better understand the molecular mechanisms of intracranial aneurysm (IA) pathogenesis, we used gene coexpression networks to identify hub genes and functional pathways associated with IA onset. Two Gene Expression Omnibus (GEO) datasets ...', 'language': 'en'}, page_content='metabolic adaptation to hypoxia. Proteins of the TP53TG3 DNA include erythropoietin, glucose transporters, glycolytic enzymes, VEGFA PROTEIN, and other factors critical to vascularization, metabolic regulation, cell multiplication and survival. Its functions annotating how cells react to oxygen were made conspicuous by the 2019 Nobel Prize in Physiology or Medicine. These pathways can interact with those resulting from GSVAB), such as the NOD2 PROTEIN signaling pathway and TP53 PROTEIN signaling pathway. We applied single-gene GSEA with the guilt by association method to annotate the functions of SLC2A12. Single-gene GSEA was usually performed after samples were divided into two groups, the highly expressed group and the lowly expressed group, according to the median expression level of the selected gene. In this GSEA, a decreasingly ranked gene list based on correlation values with SLC2A12 PROTEIN was used. Guilt by association approach in GSEA can deduce the enriched potential KEGG pathways SLC2A12 PROTEIN involved in. We deduced that SLC2A12 PROTEIN activated multiple organic substance metabolism pathways and suppressed immune response pathways including the JAK/STAT signaling pathway, CCL15 PROTEIN signaling pathway, PPIB PROTEIN and PTCRA PROTEIN signaling pathway. The roles of SLC2A12 PROTEIN in metabolism and immune response pathways merit further research. Misregulated ferroptosis can be implicated in multiple immune events23. Recently, Rui Kong reported that activation of the JAK/STAT pathway can induce ferroptosis in hepatocellular carcinoma cells24. In addition, the JAK/STAT pathway can participate in various biological processes in VSMCs25 and crosstalk with HIF1A PROTEIN pathway26,27. To date, no evidence has interpreted the correlation between SLC2A12 PROTEIN and the JAK/STAT signaling pathway. The functions of SLC2A12 PROTEIN in immune activity should be explored in the future. Based on the above literature review, SLC2A12 PROTEIN could be a novel'),\n",
       " Document(id='56958fd4-2fce-4845-a40d-3d1c799e7eb3', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/4118716/', 'title': '\\n            Involvement of the Janus Kinase/Signal Transducer and Activator of Transcription Signaling Pathway in Multiple Sclerosis and the Animal Model of Experimental Autoimmune Encephalomyelitis - PMC\\n        ', 'description': 'Multiple sclerosis (MS) and its animal model of experimental autoimmune encephalomyelitis (EAE) are characterized by focal inflammatory infiltrates into the central nervous system, demyelinating lesions, axonal damage, and abundant production of ...', 'language': 'en'}, page_content='Inc. ACTIONS View on publisher PDF DNA (460.9 KB) Cite Collections Permalink PERMALINK Copy RESOURCES Similar articles Cited by other articles Links to NCBI Databases Cite Copy Download .nbib .nbib Format: AMA APA MLA NLM Add to Collections Create a new collection Add to an existing collection Name your collection * Choose a collection Unable to load your collection due to an error Please try again Add Cancel Follow NCBI NCBI on X (formerly known as Twitter) NCBI on Facebook NCBI on LinkedIn NCBI on GitHub NCBI RSS feed Connect with NLM NLM on X (formerly known as Twitter) NLM on Facebook NLM on YouTube National Library of Medicine 8600 Rockville Pike Bethesda, MD 20894 Web Policies FOIA HHS Vulnerability Disclosure Help Accessibility Careers NLM NIH HHS USA.gov Back to Top'),\n",
       " Document(id='da38a0e1-15fb-4953-bd85-5375e90dcf4f', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/12729966/', 'title': '\\n            Integrated Transcriptomic and Proteomic Analysis Associated with Knockdown and Overexpression Studies Revealed ECHDC1 as a Regulator of Intramuscular Fat Deposition in Cattle - PMC\\n        ', 'description': 'Intramuscular fat (IMF) content is a critical determinant of beef flavor, tenderness, and juiciness. It is regulated by many factors, including nutrition, non-coding RNA, and genetic factors, among which, genetic factors play a key role. Therefore, ...', 'language': 'en'}, page_content='and Secreted by Adipocytes to Dispose of Excess Glucose, via Glycerogenesis and Increased Acyl-Glycerol Turnover. Sci. Rep. 2017;7:8983. doi: 10.1038/s41598-017-09450-4. [DOI] [PMC free article] [PubMed] [Google Scholar] 39.Kusminski C.M., Scherer P.E. Mitochondrial Dysfunction in White Adipose Tissue. Trends Endocrinol. Metab. 2012;23:435–443. doi: 10.1016/j.tem.2012.06.004. [DOI] [PMC free article] [PubMed] [Google Scholar] 40.Herzig S., Raemy E., Montessuit S., Veuthey J.-L., Zamboni N., Westermann B., Kunji E.R.S., Martinou J.-C. Identification and Functional Expression of the Mitochondrial Pyruvate Carrier. Science. 2012;337:93–96. doi: 10.1126/science.1218530. [DOI] [PubMed] [Google Scholar] 41.Gao R., Li Y., Xu Z., Zhang F., Xu J., Hu Y., Yin J., Yang K., Sun L., Wang Q., et al. Mitochondrial Pyruvate Carrier 1 Regulates Fatty Acid Synthase Lactylation and Mediates Treatment of Nonalcoholic Fatty Liver Disease. Hepatology. 2023;78:1800–1815. doi: 10.1097/HEP.0000000000000279. [DOI] [PubMed] [Google Scholar] 42.Rauckhorst A.J., Sheldon R.D., Pape D.J., Ahmed A., Falls-Hubert K.C., Merrill R.A., Brown R.F., Deshmukh K., Vallim T.A., Deja S., et al. A Hierarchical Hepatic de Novo Lipogenesis Substrate Supply Network Utilizing Pyruvate, Acetate, and Ketones. Cell Metab. 2025;37:255–273.e6. doi: 10.1016/j.cmet.2024.10.013. [DOI] [PMC free article] [PubMed] [Google Scholar] 43.Ma Y., Li Q., Chen G., Tan Z., Cao H., Bin Y., Zhou Y., Yi J., Luo X., Tan J., et al. Transcriptomic Analysis Reveals a Novel Regulatory Factor of ECHDC1 DNA Involved in Lipid Metabolism of Non-Alcoholic Fatty Liver Disease. Biochem. Biophys. Res. Commun. 2022;605:1–8. doi: 10.1016/j.bbrc.2022.03.055. [DOI] [PubMed] [Google Scholar] 44.Linster C.L., Noël G., Stroobant V., Vertommen D., Vincent M.-F., Bommer G.T., Veiga-da-Cunha M., Van Schaftingen E. Ethylmalonyl-CoA Decarboxylase, a New Enzyme Involved in Metabolite Proofreading. J. Biol. Chem. 2011;286:42992–43003. doi:'),\n",
       " Document(id='f4eb002d-6691-4fd5-ac3e-f982e1ce88be', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/11433914/', 'title': '\\n            Spastin and alsin protein interactome analyses begin to reveal key canonical pathways and suggest novel druggable targets - PMC\\n        ', 'description': 'Developing effective and long-term treatment strategies for rare and complex neurodegenerative diseases is challenging. One of the major roadblocks is the extensive heterogeneity among patients. This hinders understanding the underlying ...', 'language': 'en'}, page_content='NJ, Emr SD. The ESCRT pathway. Dev Cell. 2011;21:77–91. doi: 10.1016/j.devcel.2011.05.015. [DOI] [PubMed] [Google Scholar] Honorato RV, Koukos PI, Jimenez-Garcia B, Tsaregorodtsev A, Verlato M, Giachetti A, Rosato A, Bonvin A. Structural biology in the clouds: the WeNMR-EOSC ecosystem. Front Mol Biosci. 2021;8:729513. doi: 10.3389/fmolb.2021.729513. [DOI] [PMC free article] [PubMed] [Google Scholar] Hsu F, Spannl S, Ferguson C, Hyman AA, Parton RG, RAB5A PROTEIN and ALS2 PROTEIN regulate stress-activated cytoprotective signaling on mitochondria. Elife. 2018;7:e32282. doi: 10.7554/eLife.32282. [DOI] [PMC free article] [PubMed] [Google Scholar] Huttlin EL, et al. Dual proteome-scale networks reveal cell-specific remodeling of the human interactome. Cell. 2021;184:3022–3040. doi: 10.1016/j.cell.2021.04.011. [DOI] [PMC free article] [PubMed] [Google Scholar] Jumper J, et al. Highly accurate protein structure prediction with AlphaFold. Nature. 2021;596:583–589. doi: 10.1038/s41586-021-03819-2. [DOI] [PMC free article] [PubMed] [Google Scholar] Kanekura K, Hashimoto Y, Niikura T, Aiso S, Matsuoka M, Nishimoto I. Alsin, the product of ALS2 DNA, suppresses SOD1 PROTEIN mutant neurotoxicity through ARHGEF3 PROTEIN by interacting with SOD1 PROTEIN mutants. J Biol Chem. 2004;279:19247–19256. doi: 10.1074/jbc.M313236200. [DOI] [PubMed] [Google Scholar] Kanekura K, Hashimoto Y, Kita Y, Sasabe J, Aiso S, Nishimoto I, Matsuoka M. A Rac1/phosphatidylinositol 3-kinase/Akt3 anti-apoptotic pathway, triggered by AlsinLF, the product of the ALS2 DNA, antagonizes SOD1 PROTEIN (SOD1 PROTEIN) mutant-induced motoneuronal cell death. J Biol Chem. 2005;280:4532–4543. doi: 10.1074/jbc.M410508200. [DOI] [PubMed] [Google Scholar] Kim J, Kim S, Nahm M, Li TN, Lin HC, Kim YD, Lee J, Yao CK, Lee S. ALS2 regulates endosomal trafficking, postsynaptic development, and neuronal survival. J Cell Biol. 2021;220:e202007112. doi: 10.1083/jcb.202007112. [DOI] [PMC free article] [PubMed] [Google Scholar]'),\n",
       " Document(id='39553406-972d-4bf1-bd3f-e68a454fa34d', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/12080488/', 'title': '\\n            JAK-STAT Signaling in Autoimmunity and Cancer - PMC\\n        ', 'description': 'The JAK-STAT pathway is an essential cell survival signaling that regulates gene expressions related to inflammation, immunity and cancer. Cytokine receptors, signal transducer and activator of transcription (STAT) proteins, and Janus kinases (JAKs) ...', 'language': 'en'}, page_content='outcomes.256–258 Furthermore, STAT-III enhances UHRF1 DNA expression by directly interacting with its promoter. The PHD and SRA domains of UHRF1 PROTEIN play a crucial role in the silence of ZMYND10 DNA in colorectal cancer (CRC) cells, which is particularly significant in cancer biology.259 The characteristics of cancer, regulated by STAT-III-dependent transcriptional control of target proteins, directly affect cellular responses that promote tumor growth.260 Activation of JAK/STAT protein stimulated by IL6 PROTEIN up-regulates MMP2 PROTEIN and SNAI1 PROTEIN expression, which results in EMT.261,262 However, the JAK-II/STAT-III inhibitor WP1066 prevents IL-6– induced activation of the JAK-II/STAT-III pathway and EMT.263 Furthermore, Ovatodiolide effectively suppresses nasopharyngeal cancer development through the induction of apoptosis and inhibition of EMT, aligning with the repression of the JAK/STAT signaling pathway. The IL-6-mediated JAK/STAT signaling pathway promotes cancer progression by inhibiting autophagy.264,265 Recent studies have shown that resveratrol inhibits the IL-6–mediated JAK/STAT signaling pathway, causing autophagy and prevents ovarian cancer cell migration, according to recent studies. In primary effusion lymphoma, quercetin inhibits STAT-III to induce autophagy. Docetaxel-mediated autophagy inhibited STAT-III to reduce castration-resistant prostate cancer (CRPC) cell survival and metastasis.266–269 Another study on Head and neck squamous cell carcinomas revealed that Inhibition of either IL6R PROTEIN or EPHB2 PROTEIN effectively reverses EMT and re-sensitizes cells to ionizing radiation. According to Niu et al, the HMMR PROTEIN (HMMR PROTEIN) responsible for cell mitosis and proliferation might be implicated in the proliferation and metastasis of clear cell renal cell carcinoma (ccRCC) through EMT and JAK-I/STAT-I signaling pathways.260,270,271 Hence, many pathways like autophagy activators might be used to hinder EMT by suppressing'),\n",
       " Document(id='7a229bc6-46f6-42bc-a767-8d1f163f5350', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/7300325/', 'title': '\\n            Targeting Janus Kinases and Signal Transducer and Activator of Transcription 3 To Treat Inflammation, Fibrosis, and Cancer: Rationale, Progress, and Caution - PMC\\n        ', 'description': 'Before it was molecularly cloned in 1994, acute-phase response factor or signal transducer and activator of transcription (STAT)3 was the focus of intense research into understanding the mammalian response to injury, particularly the acute-phase ...', 'language': 'en'}, page_content='significant burden of eczematous skin disease and increased levels of IgE, prompted Siegel et al. (2013) to explore these patients for aberrant mast cell degranulation and/or reduced endothelial cell permeability. Studies in the mut-Stat3 bearing two copies of the V463del STAT3 PROTEIN mutation frequently found in Job’s syndrome patients demonstrated both defective mast cell degranulation and reduced endothelial cell permeability . In contrast, whereas human mast cell function in Job’s syndrome patients was affected somewhat by the reduced STAT3 PROTEIN signaling , the STAT3-dependent transcriptional activity within endothelial cells that regulates critical components involved in the architecture and functional dynamics of endothelial junctions, i.e., microRNA17-92, PTEN PROTEIN, SRC PROTEIN, E2F, and β-catenin, is a more important contributor to reduced endothelial cell permeability and anaphylaxis protection in these patients .B. Janus Kinase/Signal Transducer and Activator of Transcription 3 and Metabolism 1. Lipid Metabolism The anorexigenic hormone, leptin, secreted by adipose tissue, regulates energy homeostasis, glucose and lipid metabolism, immune function, and other systems . Leptin binds to its specific receptor, LepRb, on cells in the central nervous system (CNS) and peripheral tissues, including skeletal muscle, placenta, ovary, pituitary gland, and lymphoid tissue (Park and Ahima, 2014). Binding activates several intracellular signaling pathways, including JAK2 PROTEIN, STAT3 PROTEIN, IRS3P PROTEIN (IRS), PIK3CA PROTEIN, PTPN11 PROTEIN, MAPK1 PROTEIN, 5′ adenosine monophosphate-activated protein kinase, and ACACA PROTEIN. Leptin/LepRb/JAK2/STAT3 signaling plays major roles in energy homeostasis and neuroendocrine function, through upregulation of POMC DNA (POMC CELL_LINE), the precursor protein for melanocyte-stimulating hormones; corticotrophin; and β-endorphin, each of which increases appetite (Millington, 2007), and the down-regulation of'),\n",
       " Document(id='2724c922-d442-412b-be5d-4a2893535e4b', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/7300325/', 'title': '\\n            Targeting Janus Kinases and Signal Transducer and Activator of Transcription 3 To Treat Inflammation, Fibrosis, and Cancer: Rationale, Progress, and Caution - PMC\\n        ', 'description': 'Before it was molecularly cloned in 1994, acute-phase response factor or signal transducer and activator of transcription (STAT)3 was the focus of intense research into understanding the mammalian response to injury, particularly the acute-phase ...', 'language': 'en'}, page_content='injury  as well as later re-epithelialization stage of wound healing . Both pY and pS  STAT3 PROTEIN involved Persistently activated STAT3 PROTEIN, independently as well as in conjunction with TGF-β, drives unregulated wound healing, leading to fibrosis through overexpression of ECM components, e.g., COL1A2 PROTEIN , MMPs , and promoting apoptosis resistance of fibroblasts , aberrant EMT , etc. Inhibition Clinical  Ac, acetylation; AOX, ACOX2 PROTEIN; ARDS, adult respiratory distress syndrome; ATGL, adipose triglyceride lipase; C, cytoplasmic; COL1A2, COL1A2 PROTEIN; CLP, common lymphoid progenitor; DVT, deep venous thrombosis; EMT, epithelial to mesenchymal transition; FFA, free fatty acids; FLT3L, FLT3LG PROTEIN; G6PC1 PROTEIN, SLC37A4 PROTEIN; GS, genotoxic stress; HPCs, CDCA4 CELL_TYPE; iDC, immature DC; IP3, inositol 1,4,5-triphosphate; IPF, idiopathic pulmonary fibrosis; LIF, LIF PROTEIN; LPL PROTEIN, lipoprotein lipase; N, nuclear; LOF, loss of function; OS, oxidative stress; pDC, plasmacytoid DC; DLAT PROTEIN; PEPCK, PCK1 PROTEIN; PIP2, phosphatidylinositol 4,5-bisphosphate; pS, pS727; pY, pY705; RS, radiation stress; T2D, type 2 diabetes; THPO PROTEIN, THPO PROTEIN.A. Janus Kinase/Signal Transducer and Activator of Transcription 3 in Hematopoiesis and Immune Cell Function The JAK/STAT3 signaling pathway is well characterized in hematopoiesis and immune cells, where it contributes to proliferation, differentiation, activation, and regulation of multiple types of hematopoietic and immune cells in both the normal and stress conditions. 1. Granulopoiesis Although STAT3 PROTEIN was found not to be required for basal granulopoiesis , STAT3 PROTEIN was shown to regulate critical steps in emergency granulopoiesis . Specifically, CSF3 PROTEIN–mediated activation of STAT3 PROTEIN induces granulocytic progenitor proliferation through transcriptional upregulation of CEBPB PROTEIN (C/EBP)β and MYC PROTEIN . In addition, STAT3 PROTEIN activation mediates neutrophil'),\n",
       " Document(id='9513c743-3c11-4bf5-bf92-0aa5e032a488', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/12729966/', 'title': '\\n            Integrated Transcriptomic and Proteomic Analysis Associated with Knockdown and Overexpression Studies Revealed ECHDC1 as a Regulator of Intramuscular Fat Deposition in Cattle - PMC\\n        ', 'description': 'Intramuscular fat (IMF) content is a critical determinant of beef flavor, tenderness, and juiciness. It is regulated by many factors, including nutrition, non-coding RNA, and genetic factors, among which, genetic factors play a key role. Therefore, ...', 'language': 'en'}, page_content='et al. Transcriptomic Analysis Reveals a Novel Regulatory Factor of ECHDC1 DNA Involved in Lipid Metabolism of Non-Alcoholic Fatty Liver Disease. Biochem. Biophys. Res. Commun. 2022;605:1–8. doi: 10.1016/j.bbrc.2022.03.055. [DOI] [PubMed] [Google Scholar] 44.Linster C.L., Noël G., Stroobant V., Vertommen D., Vincent M.-F., Bommer G.T., Veiga-da-Cunha M., Van Schaftingen E. Ethylmalonyl-CoA Decarboxylase, a New Enzyme Involved in Metabolite Proofreading. J. Biol. Chem. 2011;286:42992–43003. doi: 10.1074/jbc.M111.281527. [DOI] [PMC free article] [PubMed] [Google Scholar] 45.Dewulf J.P., Paquay S., Marbaix E., Achouri Y., Van Schaftingen E., Bommer G.T. ECHDC1 Knockout Mice Accumulate Ethyl-Branched Lipids and Excrete Abnormal Intermediates of Branched-Chain Fatty Acid Metabolism. J. Biol. Chem. 2021;297:101083. doi: 10.1016/j.jbc.2021.101083. [DOI] [PMC free article] [PubMed] [Google Scholar] 46.Althaher A.R. An Overview of LIPE PROTEIN (HSL) Sci. World J. 2022;2022:1964684. doi: 10.1155/2022/1964684. [DOI] [PMC free article] [PubMed] [Google Scholar] Associated Data This section collects any data citations, data availability statements, or supplementary materials included in this article. Data Availability Statement The datasets used and/or analyzed during the current study are available from the corresponding author upon reasonable request. Articles from Animals : an Open Access Journal from MDPI are provided here courtesy of Multidisciplinary Digital Publishing Institute (MDPI) ACTIONS View on publisher PDF DNA (2.2 MB) Cite Collections Permalink PERMALINK Copy RESOURCES Similar articles Cited by other articles Links to NCBI Databases Cite Copy Download .nbib .nbib Format: AMA APA MLA NLM Add to Collections Create a new collection Add to an existing collection Name your collection * Choose a collection Unable to load your collection due to an error Please try again Add Cancel Follow NCBI NCBI on X (formerly known as Twitter) NCBI on Facebook NCBI on LinkedIn NCBI'),\n",
       " Document(id='c0070964-55de-4591-b1b5-ffa4004fc8fb', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/11433914/', 'title': '\\n            Spastin and alsin protein interactome analyses begin to reveal key canonical pathways and suggest novel druggable targets - PMC\\n        ', 'description': 'Developing effective and long-term treatment strategies for rare and complex neurodegenerative diseases is challenging. One of the major roadblocks is the extensive heterogeneity among patients. This hinders understanding the underlying ...', 'language': 'en'}, page_content='with the proteins that interact with distinct domains. We found that the full-length protein was part of key cellular events with P-values suggesting high significance (ranging from 1.18E-09 to 5.21E-16) and that each domain was particularly involved in a select set of cellular events, some of which were shared and some were unique. An overall look at the canonical pathways suggested that the endosomal sorting complex required for transport (ESCRT) was one of the key cellular activities for spastin protein, where the full-length protein (P = 5.21E-16), HD domain (P = 5.41E-07), MIT domain (P = 9.51E-8) and the AAA domain (P = 6.29E-08) all participated with significance that cannot be explained by luck. The ESCRT pathway plays a major role in the formation of endosomal compartments and vesicular trafficking . It consists of several complexes, including ESCRT-0, -I, -II, -III, and the VPS4A PROTEIN, each playing a distinct role in the formation of intraluminal vesicles and multi-vesicular bodies. The connection between the spastin protein and ESCRT proteins has been previously reported . ESCRT proteins not only help with this localization, but they also assist in polarized membrane trafficking, coupled with the enzymatic activity of the AAA region . Additionally, dysfunction of the ESCRT pathway has been canonically reported to have neurodegenerative effects, especially through stress-mediated pathways . Therefore, the relationship between ESCRT and the spastin protein interactome is robust and offers novel mechanistic insight into a potential avenue of pathogenesis. Several of the other identified canonical pathways in the spastin protein interactome demonstrated both a functional and mechanistic relationship with the ESCRT pathway. For example, autophagy, which is significantly related to the spastin protein interactome through the late endosomal microautophagy and macroautophagy pathways, is mediated through components of ESCRT-I and ESCRT-III . The late'),\n",
       " Document(id='d95cd45d-a8c9-4cc3-94ee-ce51059e98a1', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/12729966/', 'title': '\\n            Integrated Transcriptomic and Proteomic Analysis Associated with Knockdown and Overexpression Studies Revealed ECHDC1 as a Regulator of Intramuscular Fat Deposition in Cattle - PMC\\n        ', 'description': 'Intramuscular fat (IMF) content is a critical determinant of beef flavor, tenderness, and juiciness. It is regulated by many factors, including nutrition, non-coding RNA, and genetic factors, among which, genetic factors play a key role. Therefore, ...', 'language': 'en'}, page_content='Eagle Medium/F12 Nutrient Mixture MSC mesenchymal stem cell ECHDC1 Enoyl-CoA Hydratase Domain-Containing Protein 1 HSL hormone-sensitive lipase FABP4 fatty acid binding protein 4 PPARγ PPARA PROTEIN γ C/EBPα CCAAT/enhancer binding protein ACACA Acetyl-CoA carboxylase Alpha ATGL adipose triglyceride lipase SREBP-1 Sterol Regulatory Element-Binding Protein-1 SNCG Synuclein Gamma ECHDC1 Enoyl-CoA Hydratase Domain Containing 1 FGF1 Fibroblast Growth Factor 1 CLEC11A C-Type Lectin Domain Family 11 Member A PHYHIPL Phytanoyl-CoA Hydroxylase Interacting Protein-Like RIDA RAD51 Inhibitor Domain-Containing Protein ACSS2 Acyl-CoA Synthetase Short-Chain Family Member 2 QPRT PROTEIN Quinolinate Phosphoribosyltransferase CDH2 Cadherin 2 (N-Cadherin) COL3A1 Collagen Type III Alpha 1 Chain H2AC11 H2AC20 PROTEIN Clustered H4C12 PROTEIN GSTA6P PROTEIN RGN Regucalcin TMSB15A Thymosin Beta 15A FIBIN Fibulin 1 PPP1R12B Protein Phosphatase 1 Regulatory Subunit 12B PDK4 Pyruvate Dehydrogenase Kinase 4 IFITM1 PROTEIN Interferon-Induced Transmembrane Protein 1 CFH Complement Factor H MYH6 Myosin Heavy Chain 6 (Alpha-Myosin) mTOR mammalian target of rapamycin MAPK1 PROTEIN Mitogen-Activated Protein Kinase Wnt Wingless-Type MMTV Integration Site Family JAK-STAT Janus Kinase-Signal Transducer and Activator of Transcription TPM Transcripts Per Million FDR false discovery rate FC fold change GO Gene Ontology KEGG Kyoto Encyclopedia of Genes and Genomes LC-MS/MS liquid chromatography-tandem mass spectrometry DIA Data-Independent Acquisition RT-qPCR Real-time Fluorescence Quantitative PCR PBS Phosphate-Buffered Saline SDS-PAGE Sodium Dodecyl Sulfate-Polyacrylamide Gel Electrophoresis BCA Bicinchoninic Acid Assay Supplementary Materials The following supporting information can be downloaded at: https://doi.org/10.5281/zenodo.17242096. Figure S1: Full Scans of the Western Blots; Table S1: Antibody information.Author Contributions Conceptualization, W.Y.; methodology, R.H.; validation, R.H., L.L.,'),\n",
       " Document(id='9129bf62-f6c2-4f0a-b96b-4a807cedee88', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/7327787/', 'title': '\\n            Interventions of natural and synthetic agents in inflammatory bowel disease, modulation of nitric oxide pathways - PMC\\n        ', 'description': 'Inflammatory bowel disease (IBD) refers to a group of disorders characterized by chronic inflammation of the gastrointestinal (GI) tract. The elevated levels of nitric oxide (NO) in serum and affected tissues; mainly synthesized by the inducible ...', 'language': 'en'}, page_content='are provided here courtesy of Baishideng Publishing Group Inc ACTIONS View on publisher PDF DNA (2.4 MB) Cite Collections Permalink PERMALINK Copy RESOURCES Similar articles Cited by other articles Links to NCBI Databases Cite Copy Download .nbib .nbib Format: AMA APA MLA NLM Add to Collections Create a new collection Add to an existing collection Name your collection * Choose a collection Unable to load your collection due to an error Please try again Add Cancel Follow NCBI NCBI on X (formerly known as Twitter) NCBI on Facebook NCBI on LinkedIn NCBI on GitHub NCBI RSS feed Connect with NLM NLM on X (formerly known as Twitter) NLM on Facebook NLM on YouTube National Library of Medicine 8600 Rockville Pike Bethesda, MD 20894 Web Policies FOIA HHS Vulnerability Disclosure Help Accessibility Careers NLM NIH HHS USA.gov Back to Top'),\n",
       " Document(id='d7d3bf01-dd19-485d-9145-89ac55f5e791', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/12729966/', 'title': '\\n            Integrated Transcriptomic and Proteomic Analysis Associated with Knockdown and Overexpression Studies Revealed ECHDC1 as a Regulator of Intramuscular Fat Deposition in Cattle - PMC\\n        ', 'description': 'Intramuscular fat (IMF) content is a critical determinant of beef flavor, tenderness, and juiciness. It is regulated by many factors, including nutrition, non-coding RNA, and genetic factors, among which, genetic factors play a key role. Therefore, ...', 'language': 'en'}, page_content='the mitochondria, thereby affecting pyruvate metabolism  Under aerobic conditions, pyruvate enters the mitochondria through the MPC and undergoes decarboxylation under the catalysis of the DLAT PROTEIN to form acetyl-CoA, which then enters the tricarboxylic acid cycle  Research has found that interference with MPC can suppress the ability to synthesize lipids and glycerol-3-phosphate, leading to a reduction in triglyceride accumulation  In addition, pyruvate can also synthesize SLC33A1 PROTEIN through the mitochondrial-cytoplasmic MPC–ATP–citrate lyase pathway, initiating de novo synthesis of cytoplasmic fatty acids  The aforementioned studies indicate that pyruvate metabolism plays a significant role in fat deposition. This study found through combined analysis that ECHDC1 DNA was differentially expressed in intramuscular fat tissues of beef with varying grades, and previous studies found that interference with this gene inhibited the number of intracellular lipid droplets  However, existing studies have not clearly explored the regulatory role of the ECHDC1 DNA in the adipogenesis of bovine intramuscular adipocytes, and the potential interaction between ECHDC1 DNA and other lipid metabolism-related genes remains to be further clarified. Based on the above research gaps, this study hypothesizes that the ECHDC1 DNA may play a key regulatory role in bovine intramuscular fat deposition and intends to initially investigate its molecular regulatory mechanisms. ECHDC1 DNA is a cytoplasmic enzyme that catalyzes the decarboxylation of ethyl or methyl malonyl-CoA  thereby playing a role in fatty acid synthesis and restricting the synthesis of abnormal fatty acids, such as ethyl or methyl-branched fatty acids  This study analyzed the spatiotemporal specificity of ECHDC1 DNA and found that it may be involved in the early stages of preadipocyte differentiation. Its high expression in the longissimus dorsi muscle tissue further suggests that it may play a role in intramuscular'),\n",
       " Document(id='3772a69d-d64a-43cc-a2ac-bae460dee8c9', metadata={'source': 'https://pmc.ncbi.nlm.nih.gov/articles/11433914/', 'title': '\\n            Spastin and alsin protein interactome analyses begin to reveal key canonical pathways and suggest novel druggable targets - PMC\\n        ', 'description': 'Developing effective and long-term treatment strategies for rare and complex neurodegenerative diseases is challenging. One of the major roadblocks is the extensive heterogeneity among patients. This hinders understanding the underlying ...', 'language': 'en'}, page_content='M, Liu WJ, Qiu Y; T-Editor: Jia Y Data availability statement: All relevant data are within the paper and its Additional files.References Allison R, Lumb JH, Fassier C, Connell JW, Ten Martin D, Seaman MN, Hazan J, Reid E. An ESCRT-spastin interaction promotes fission of recycling tubules from the endosome. J Cell Biol. 2013;202:527–543. doi: 10.1083/jcb.201211045. [DOI] [PMC free article] [PubMed] [Google Scholar] Allison R, Edgar JR, Pearson G, Rizo T, Newton T, Gunther S, Berner F, Hague J, Connell JW, Winkler J, Lippincott-Schwartz J, Beetz C, Winner B, Reid E. Defects in ER-endosome contacts impact lysosome function in hereditary spastic paraplegia. J Cell Biol. 2017;216:1337–1355. doi: 10.1083/jcb.201609033. [DOI] [PMC free article] [PubMed] [Google Scholar] Allison R, Edgar JR, Reid E. Spastin MIT domain disease-associated mutations disrupt lysosomal function. Front Neurosci. 2019;13:1179. doi: 10.3389/fnins.2019.01179. [DOI] [PMC free article] [PubMed] [Google Scholar] Arribat Y, Grepper D, Lagarrigue S, Qi T, Cohen S, Amati F. Spastin mutations impair coordination between lipid droplet dispersion and reticulum. PLoS Genet. 2020;16:e1008665. doi: 10.1371/journal.pgen.1008665. [DOI] [PMC free article] [PubMed] [Google Scholar] Barone C, Qi X. Altered metabolism in motor neuron diseases: mechanism and potential therapeutic target. Cells. 2023;12:1536. doi: 10.3390/cells12111536. [DOI] [PMC free article] [PubMed] [Google Scholar] Brewer PD, Habtemichael EN, Romenskaia I, Mastick CC, Coster AC. Insulin-regulated SLC2A4 PROTEIN translocation: TMEM50A PROTEIN trafficking with six distinctive steps. J Biol Chem. 2014;289:17280–17298. doi: 10.1074/jbc.M114.555714. [DOI] [PMC free article] [PubMed] [Google Scholar] Bryant P, Pozzati G, Elofsson A. Improved prediction of protein-protein interactions using AlphaFold2. Nat Commun. 2022;13:1265. doi: 10.1038/s41467-022-28865-w. [DOI] [PMC free article] [PubMed] [Google Scholar] Cannariato M, Miceli M, Deriu MA. In')]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "retriever.invoke('explain details EMT and which pathways are involved in this process?')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "172af392",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "aplay: pcm_write:2146: write error: Interrupted system call\n"
     ]
    }
   ],
   "source": [
    "engine = pyttsx3.init()\n",
    "output='Usage'\n",
    "text = f\"The output of Large Language model: {response}\"\n",
    "engine.setProperty('rate', 160)\n",
    "engine.setProperty('volume',0.7)\n",
    "voices = engine.getProperty('voices')\n",
    "engine.setProperty('voice', voices[23].id)\n",
    "engine.say(text)\n",
    "engine.runAndWait()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "103791bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "docs_content = \"\\n\\n\".join(doc.page_content for doc in vectorstore.similarity_search('which Signalling pathways are involved in human diseases'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "24a79a8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The GHR PROTEIN signaling pathway plays important roles in growth, metabolism, cell cycle control, immunity, homeostatic processes, and chemoresistance via both the JAK/STAT and the SRC PROTEIN pathways. Dysregulation of GHR PROTEIN signaling is associated with various diseases and chronic conditions such as acromegaly, cancer, aging, metabolic disease, fibroses, inflammation and autoimmunity. Numerous studies entailing the GHR PROTEIN signaling pathway have been conducted for various cancers. Diverse factors mediate the up- or down-regulation of GHR PROTEIN signaling through post-translational modifications. Of the numerous modifications, ubiquitination and deubiquitination are prominent events. Ubiquitination by PRKN PROTEIN attaches UBD PROTEIN to target proteins and induces proteasomal degradation or starts the sequence of events that leads to endocytosis and lysosomal degradation. In this review, we discuss the role of first line effectors that act directly on the GHR PROTEIN at the cell surface including ADAM17 PROTEIN, JAK2 PROTEIN, SRC family member Lyn, Ubc13/CHIP, proteasome, βTrCP, CK2, STAT5B PROTEIN, and SOCS2 PROTEIN. Activity of all, except JAK2 PROTEIN, LYN PROTEIN and STAT5B PROTEIN, counteract GHR PROTEIN signaling. Loss of their function increases the GH-induced signaling in favor of aging and certain chronic diseases, exemplified by increased lung cancer risk in case of a mutation in the SOCS2-GHR interaction site. Insight in their roles in GHR PROTEIN signaling can be applied for cancer and other therapeutic strategies. : aging, cancer, ubiquitin, endocytosis, POU1F1 PROTEIN axis, GH sensitivityIntroduction In 1989 with a background of posttranslational modifications and intracellular transport of IL6R PROTEIN our lab decided to focus on studying the role of ubiquitination in membrane trafficking. Knowledge on the role of ubiquitination as a major regulator of cell functions had just started to emerge . To address the question of whether\\n\\nby GH was shown to activate MAPKs, also referred to as MAPK1 PROTEIN (MAPK3 PROTEIN and MAPK1 PROTEIN), through the PLCD1 PROTEIN Cγ-Ras pathway, signaling that might promote oncogenesis . Genes exclusively regulated by LYN PROTEIN include genes involved in mRNA transcription and metabolism, including the GHR PROTEIN itself: the basal GHR PROTEIN expression level via LYN PROTEIN is 4.8-fold higher, comparing GHR Box1-/- vs. GHR-/- . GHR signaling via this pathway induces also HLA-G, a powerful immunosuppressive protein for NK cells and macrophages. GH-enhanced immunosuppression in tumors might evade immune attack. On the other hand, it might be used to stop excessive inflammation after partial hepatectomy allowing liver regeneration and survival, Figs. 4 and 7 . For 3T3-F442A preadipocytes and H4IIE hepatoma cells it has been shown that relative levels of JAK2 PROTEIN and SRC family kinase in any particular cell might determine which MAP2K1 PROTEIN is the major signaling element, with JAK2 PROTEIN predominating in most cases . Barclay and co-workers showed that targeted mutation in the STUB1 PROTEIN of GHR PROTEIN in mice, although abrogating JAK2 PROTEIN activation, did not decrease the hepatic activation of MAPK1 PROTEIN via LYN PROTEIN . The importance of this pathway came from studies with exon 3-deleted GHR, which results in the deletion of 22 amino acids in the LYVE1 PROTEIN of the GHR. Males with this genotype exhibit reduced basal but enhanced EPHB2 PROTEIN signaling after GH stimulation. Exon 3-deleted GHR individuals showed lower serum IGF-1 levels, and were found to be of higher stature with extended lifespan (10 years) .Signaling Pathways of GH1 PROTEIN The main pathways activated by GH are: the STAT1 PROTEIN pathway, the MAPKAPK2 PROTEIN (MAPK1 PROTEIN) pathway, and the PIK3CD PROTEIN (PIK3CA PROTEIN) pathway. The extent by which each pathway is activated depends on the cell types, related to differences in relative expression levels of the components\\n\\nby GH was shown to activate MAPKs, also referred to as MAPK1 PROTEIN (MAPK3 PROTEIN and MAPK1 PROTEIN), through the PLCD1 PROTEIN Cγ-Ras pathway, signaling that might promote oncogenesis . Genes exclusively regulated by LYN PROTEIN include genes involved in mRNA transcription and metabolism, including the GHR PROTEIN itself: the basal GHR PROTEIN expression level via LYN PROTEIN is 4.8-fold higher, comparing GHR Box1-/- vs. GHR-/- . GHR signaling via this pathway induces also HLA-G, a powerful immunosuppressive protein for NK cells and macrophages. GH-enhanced immunosuppression in tumors might evade immune attack. On the other hand, it might be used to stop excessive inflammation after partial hepatectomy allowing liver regeneration and survival, Figs. 4 and 7 . For 3T3-F442A preadipocytes and H4IIE hepatoma cells it has been shown that relative levels of JAK2 PROTEIN and SRC family kinase in any particular cell might determine which MAP2K1 PROTEIN is the major signaling element, with JAK2 PROTEIN predominating in most cases . Barclay and co-workers showed that targeted mutation in the STUB1 PROTEIN of GHR PROTEIN in mice, although abrogating JAK2 PROTEIN activation, did not decrease the hepatic activation of MAPK1 PROTEIN via LYN PROTEIN . The importance of this pathway came from studies with exon 3-deleted GHR, which results in the deletion of 22 amino acids in the LYVE1 PROTEIN of the GHR. Males with this genotype exhibit reduced basal but enhanced EPHB2 PROTEIN signaling after GH stimulation. Exon 3-deleted GHR individuals showed lower serum IGF-1 levels, and were found to be of higher stature with extended lifespan (10 years) .Signaling Pathways of GH1 PROTEIN The main pathways activated by GH are: the STAT1 PROTEIN pathway, the MAPKAPK2 PROTEIN (MAPK1 PROTEIN) pathway, and the PIK3CD PROTEIN (PIK3CA PROTEIN) pathway. The extent by which each pathway is activated depends on the cell types, related to differences in relative expression levels of the components\\n\\npathway. In the scheme also molecules involved in signal termination (PTS PROTEIN, CISH PROTEIN and PIAS1 PROTEIN) are indicated. Inthe physiology and pathophysiology are further detailed.The phosphorylation of specific tyrosine residues, brought about by JAK2 PROTEIN and LYN PROTEIN, has extensively been studied as intermediaries for recruiting downstream signaling effectors . Herewith in agreement, we confirmed that tyrosine residues 487, 534, and 627, but not residue 566, are most important for GHR PROTEIN and STAT5B PROTEIN phosphorylation. In addition, we showed that the GHR(Y627F) mutation constitutively (independent of GH binding) activates JAK2 and downstream effectors .JAK2 Activation JAK2 PROTEIN is composed of four major domains: a N-terminal FERM (4.1 protein, EZR PROTEIN, RDXP1 PROTEIN, and moesin) domain, followed by a SH2 (Src homology 2) domain, pseudokinase and kinase domains. The binding to the STUB1 PROTEIN of GHR PROTEIN occurs through the FRMD4A PROTEIN . Normally JAK2 PROTEIN is held in an inactive conformation, where the MAP2K1 PROTEIN and pseudokinase domain interact with each other . The activation of JAK2 PROTEIN requires that two catalytic domains are brought in close proximity. This was concluded after realizing that often in human leukemia there is oligomerization of JAK2 molecules which renders them constitutively active; this aggregation is due to the occurrence of a genetic fusion between the JAK2 catalytic domain and the oncogenic transcription factor TEL . Analysis of other mutation also contributed to the understanding of JAK2 PROTEIN physiology. The mutation V617A, which turns JAK2 PROTEIN into a constitutively active state, is found in patients with myeloproliferative disorders . This mutation probably disrupts the inhibitory interaction between the pseudokinase and kinase domain . Following mutational analysis, the SH2-pseudokinase domain linker turned out to be important for JAK2 PROTEIN activation . JAK2 PROTEIN activation'"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "docs_content"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LLM",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
